// @delon/abc/__ivy_ngcc__/fesm2015/table.js
import * as i0 from '@angular/core';
import { Injectable, Directive, TemplateRef, Host, Input, Optional, Inject, ViewContainerRef, ComponentFactoryResolver, EventEmitter, Component, ChangeDetectionStrategy, ViewEncapsulation, ChangeDetectorRef, ElementRef, ViewChild, Output, NgModule } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { ACLService, DelonACLModule } from '@delon/acl';
import { ALAIN_I18N_TOKEN, _HttpClient, DatePipe, YNPipe, DelonLocaleService, ModalHelper, DrawerHelper } from '@delon/theme';
import { warn, deepCopy, deepGet, deepMergeKey } from '@delon/util/other';
import { DecimalPipe, DOCUMENT, CommonModule } from '@angular/common';
import { HttpParams } from '@angular/common/http';
import { of, Subject, from, isObservable } from 'rxjs';
import { map, takeUntil, filter } from 'rxjs/operators';
import { CurrencyService } from '@delon/util/format';
import { __awaiter, __decorate } from 'tslib';
import { XlsxService } from '@delon/abc/xlsx';
import { AlainConfigService } from '@delon/util/config';
import { toBoolean, InputNumber, InputBoolean } from '@delon/util/decorator';
import { NzContextMenuService, NzDropDownModule } from 'ng-zorro-antd/dropdown';
import { FormsModule } from '@angular/forms';
import { LetModule } from '@delon/abc/let';
import { NzBadgeModule } from 'ng-zorro-antd/badge';
import { NzCheckboxModule } from 'ng-zorro-antd/checkbox';
import { NzDatePickerModule } from 'ng-zorro-antd/date-picker';
import { NzDividerModule } from 'ng-zorro-antd/divider';
import { NzIconModule } from 'ng-zorro-antd/icon';
import { NzInputModule } from 'ng-zorro-antd/input';
import { NzInputNumberModule } from 'ng-zorro-antd/input-number';
import { NzMenuModule } from 'ng-zorro-antd/menu';
import { NzPopconfirmModule } from 'ng-zorro-antd/popconfirm';
import { NzRadioModule } from 'ng-zorro-antd/radio';
import { NzResizableModule } from 'ng-zorro-antd/resizable';
import { NzTableModule } from 'ng-zorro-antd/table';
import { NzTagModule } from 'ng-zorro-antd/tag';
import { NzToolTipModule } from 'ng-zorro-antd/tooltip';
import { Router } from '@angular/router';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/platform-browser';
import * as ɵngcc2 from '@delon/acl';
import * as ɵngcc3 from '@delon/theme';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@delon/util/format';
import * as ɵngcc6 from '@delon/abc/xlsx';
import * as ɵngcc7 from '@delon/util/config';
import * as ɵngcc8 from 'ng-zorro-antd/dropdown';
import * as ɵngcc9 from 'ng-zorro-antd/table';
import * as ɵngcc10 from 'ng-zorro-antd/menu';
import * as ɵngcc11 from 'ng-zorro-antd/tooltip';
import * as ɵngcc12 from 'ng-zorro-antd/icon';
import * as ɵngcc13 from 'ng-zorro-antd/checkbox';
import * as ɵngcc14 from '@angular/forms';
import * as ɵngcc15 from '@delon/abc/let';
import * as ɵngcc16 from 'ng-zorro-antd/resizable';
import * as ɵngcc17 from 'ng-zorro-antd/input';
import * as ɵngcc18 from 'ng-zorro-antd/input-number';
import * as ɵngcc19 from 'ng-zorro-antd/date-picker';
import * as ɵngcc20 from 'ng-zorro-antd/radio';
import * as ɵngcc21 from '@angular/router';
import * as ɵngcc22 from 'ng-zorro-antd/popconfirm';
import * as ɵngcc23 from 'ng-zorro-antd/tag';
import * as ɵngcc24 from 'ng-zorro-antd/badge';
import * as ɵngcc25 from 'ng-zorro-antd/divider';

const _c0 = ["table"];
const _c1 = ["contextmenuTpl"];
function STComponent_ng_template_0_small_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "small", 14);
} if (rf & 2) {
    const i_r16 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", i_r16.optional, ɵngcc0.ɵɵsanitizeHtml);
} }
function STComponent_ng_template_0_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 15);
} if (rf & 2) {
    const i_r16 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("nzTooltipTitle", i_r16.optionalHelp);
} }
function STComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 11);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_template_0_small_1_Template, 1, 1, "small", 12);
    ɵngcc0.ɵɵtemplate(2, STComponent_ng_template_0_i_2_Template, 1, 1, "i", 13);
} if (rf & 2) {
    const i_r16 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", i_r16._text, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r16.optional);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r16.optionalHelp);
} }
function STComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    const _r23 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 16);
    ɵngcc0.ɵɵlistener("ngModelChange", function STComponent_ng_template_2_Template_label_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r23); const ctx_r22 = ɵngcc0.ɵɵnextContext(); return ctx_r22._allChecked = $event; })("ngModelChange", function STComponent_ng_template_2_Template_label_ngModelChange_0_listener() { ɵngcc0.ɵɵrestoreView(_r23); const ctx_r24 = ɵngcc0.ɵɵnextContext(); return ctx_r24._checkAll(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const custom_r21 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("ant-table-selection-select-all-custom", custom_r21);
    ɵngcc0.ɵɵproperty("nzDisabled", ctx_r3._allCheckedDisabled)("ngModel", ctx_r3._allChecked)("nzIndeterminate", ctx_r3._indeterminate);
} }
function STComponent_thead_6_tr_1_th_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "th", 18);
} if (rf & 2) {
    const ctx_r28 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("rowSpan", ctx_r28._headers.length);
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_nz_resize_handle_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "nz-resize-handle", 25);
    ɵngcc0.ɵɵelement(1, "i");
    ɵngcc0.ɵɵelementEnd();
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_template_2_Template(rf, ctx) { }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_ng_container_1_ng_template_1_Template(rf, ctx) { }
const _c2 = function () { return { $implicit: false }; };
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext(7);
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction0(2, _c2));
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_ng_template_1_Template(rf, ctx) { }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 35);
    ɵngcc0.ɵɵelementStart(1, "div", 36);
    ɵngcc0.ɵɵelement(2, "i", 37);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r47 = ɵngcc0.ɵɵreference(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("nzDropdownMenu", _r47);
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_li_6_Template(rf, ctx) { if (rf & 1) {
    const _r51 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 38);
    ɵngcc0.ɵɵlistener("click", function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_li_6_Template_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r51); const rw_r49 = restoredCtx.$implicit; const ctx_r50 = ɵngcc0.ɵɵnextContext(8); return ctx_r50._rowSelection(rw_r49); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const rw_r49 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", rw_r49.text, ɵngcc0.ɵɵsanitizeHtml);
} }
const _c3 = function () { return { $implicit: true }; };
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 30);
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_ng_template_1_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵtemplate(2, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_div_2_Template, 3, 1, "div", 31);
    ɵngcc0.ɵɵelementStart(3, "nz-dropdown-menu", null, 32);
    ɵngcc0.ɵɵelementStart(5, "ul", 33);
    ɵngcc0.ɵɵtemplate(6, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_li_6_Template, 1, 1, "li", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _c_r34 = ɵngcc0.ɵɵnextContext(3).let;
    ɵngcc0.ɵɵnextContext(4);
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction0(4, _c3));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", _c_r34.selections.length);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngForOf", _c_r34.selections);
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_ng_container_1_Template, 2, 3, "ng-container", 4);
    ɵngcc0.ɵɵtemplate(2, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_div_2_Template, 7, 5, "div", 29);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _c_r34 = ɵngcc0.ɵɵnextContext(2).let;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", _c_r34.selections.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", _c_r34.selections.length > 0);
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_3_ng_template_1_Template(rf, ctx) { }
const _c4 = function (a0) { return { $implicit: a0 }; };
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_3_ng_template_1_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _c_r34 = ɵngcc0.ɵɵnextContext(2).let;
    ɵngcc0.ɵɵnextContext(4);
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c4, _c_r34.title));
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementContainerStart(1, 26);
    ɵngcc0.ɵɵtemplate(2, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_2_Template, 3, 2, "ng-container", 27);
    ɵngcc0.ɵɵtemplate(3, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_ng_container_3_Template, 2, 4, "ng-container", 28);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _c_r34 = ɵngcc0.ɵɵnextContext().let;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitch", _c_r34.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "checkbox");
} }
function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    const _r58 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "st-filter", 39);
    ɵngcc0.ɵɵlistener("n", function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_5_Template_st_filter_n_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r58); const ctx_r57 = ɵngcc0.ɵɵnextContext(5); return ctx_r57.handleFilterNotify($event); })("handle", function STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_5_Template_st_filter_handle_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r58); const _c_r34 = ɵngcc0.ɵɵnextContext().let; const ctx_r59 = ɵngcc0.ɵɵnextContext(4); return ctx_r59._handleFilter(_c_r34, $event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _c_r34 = ɵngcc0.ɵɵnextContext().let;
    const h_r30 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r39 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("col", h_r30.column)("f", _c_r34.filter)("locale", ctx_r39.locale);
} }
const _c5 = function (a0, a1) { return { $implicit: a0, index: a1 }; };
function STComponent_thead_6_tr_1_ng_container_2_th_1_Template(rf, ctx) { if (rf & 1) {
    const _r64 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "th", 20);
    ɵngcc0.ɵɵlistener("nzSortOrderChange", function STComponent_thead_6_tr_1_ng_container_2_th_1_Template_th_nzSortOrderChange_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r64); const _c_r34 = restoredCtx.let; const index_r31 = ɵngcc0.ɵɵnextContext().index; const ctx_r63 = ɵngcc0.ɵɵnextContext(3); return ctx_r63.sort(_c_r34, index_r31, $event); })("nzResizeEnd", function STComponent_thead_6_tr_1_ng_container_2_th_1_Template_th_nzResizeEnd_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r64); const _c_r34 = restoredCtx.let; const ctx_r66 = ɵngcc0.ɵɵnextContext(4); return ctx_r66.colResize($event, _c_r34); });
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_ng_container_2_th_1_nz_resize_handle_1_Template, 2, 0, "nz-resize-handle", 21);
    ɵngcc0.ɵɵtemplate(2, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_template_2_Template, 0, 0, "ng-template", 22, 23, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(4, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_4_Template, 4, 2, "ng-container", 24);
    ɵngcc0.ɵɵtemplate(5, STComponent_thead_6_tr_1_ng_container_2_th_1_ng_container_5_Template, 2, 3, "ng-container", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _c_r34 = ctx.let;
    const _r36 = ɵngcc0.ɵɵreference(3);
    const ctx_r67 = ɵngcc0.ɵɵnextContext();
    const h_r30 = ctx_r67.$implicit;
    const last_r32 = ctx_r67.last;
    const index_r31 = ctx_r67.index;
    ɵngcc0.ɵɵproperty("colSpan", h_r30.colSpan)("rowSpan", h_r30.rowSpan)("nzWidth", _c_r34.width)("nzLeft", _c_r34._left)("nzRight", _c_r34._right)("ngClass", _c_r34.className)("nzShowSort", _c_r34._sort.enabled)("nzSortOrder", _c_r34._sort.default)("nzCustomFilter", _c_r34.filter)("nzDisabled", last_r32 || _c_r34.resizable.disabled)("nzMaxWidth", _c_r34.resizable.maxWidth)("nzMinWidth", _c_r34.resizable.minWidth)("nzBounds", _c_r34.resizable.bounds)("nzPreview", _c_r34.resizable.preview);
    ɵngcc0.ɵɵattribute("data-col", _c_r34.indexKey)("data-col-index", index_r31);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !last_r32 && !_c_r34.resizable.disabled);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _c_r34.__renderTitle)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(22, _c5, h_r30.column, index_r31));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !_c_r34.__renderTitle)("ngIfElse", _r36);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", _c_r34.filter);
} }
function STComponent_thead_6_tr_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_ng_container_2_th_1_Template, 6, 25, "th", 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const h_r30 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("let", h_r30.column);
} }
function STComponent_thead_6_tr_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "tr");
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_th_1_Template, 1, 1, "th", 17);
    ɵngcc0.ɵɵtemplate(2, STComponent_thead_6_tr_1_ng_container_2_Template, 2, 1, "ng-container", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const row_r26 = ctx.$implicit;
    const rowFirst_r27 = ctx.first;
    const ctx_r25 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", rowFirst_r27 && ctx_r25.expand);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", row_r26);
} }
function STComponent_thead_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "thead");
    ɵngcc0.ɵɵtemplate(1, STComponent_thead_6_tr_1_Template, 3, 2, "tr", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r5._headers);
} }
function STComponent_ng_container_8_ng_template_1_Template(rf, ctx) { }
function STComponent_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_container_8_ng_template_1_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r6.bodyHeader)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c4, ctx_r6._statistical));
} }
function STComponent_ng_template_9_td_1_Template(rf, ctx) { if (rf & 1) {
    const _r76 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "td", 44);
    ɵngcc0.ɵɵlistener("nzExpandChange", function STComponent_ng_template_9_td_1_Template_td_nzExpandChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r76); const i_r69 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r74 = ɵngcc0.ɵɵnextContext(); return ctx_r74._expandChange(i_r69, $event); })("click", function STComponent_ng_template_9_td_1_Template_td_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r76); const ctx_r77 = ɵngcc0.ɵɵnextContext(2); return ctx_r77._stopPropagation($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r69 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r71 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("nzShowExpand", ctx_r71.expand && i_r69.showExpand !== false)("nzExpand", i_r69.expand);
} }
function STComponent_ng_template_9_td_2_span_1_ng_template_1_Template(rf, ctx) { }
function STComponent_ng_template_9_td_2_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 48);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_template_9_td_2_span_1_ng_template_1_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const c_r79 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext(2);
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c4, c_r79.title));
} }
function STComponent_ng_template_9_td_2_Template(rf, ctx) { if (rf & 1) {
    const _r85 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "td", 45);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_template_9_td_2_span_1_Template, 2, 4, "span", 46);
    ɵngcc0.ɵɵelementStart(2, "st-td", 47);
    ɵngcc0.ɵɵlistener("n", function STComponent_ng_template_9_td_2_Template_st_td_n_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r85); const ctx_r84 = ɵngcc0.ɵɵnextContext(2); return ctx_r84._handleTd($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const c_r79 = ctx.$implicit;
    const cIdx_r80 = ctx.index;
    const ctx_r86 = ɵngcc0.ɵɵnextContext();
    const i_r69 = ctx_r86.$implicit;
    const index_r70 = ctx_r86.index;
    const ctx_r72 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("nzLeft", !!c_r79._left)("nzRight", !!c_r79._right)("ngClass", c_r79._className);
    ɵngcc0.ɵɵattribute("data-col-index", cIdx_r80)("colspan", c_r79.colSpan);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r72.responsive);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("data", ctx_r72._data)("i", i_r69)("index", index_r70)("c", c_r79)("cIdx", cIdx_r80);
} }
function STComponent_ng_template_9_ng_template_4_Template(rf, ctx) { }
function STComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    const _r88 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 40);
    ɵngcc0.ɵɵlistener("click", function STComponent_ng_template_9_Template_tr_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r88); const i_r69 = restoredCtx.$implicit; const index_r70 = restoredCtx.index; const ctx_r87 = ɵngcc0.ɵɵnextContext(); return ctx_r87._rowClick($event, i_r69, index_r70, false); })("dblclick", function STComponent_ng_template_9_Template_tr_dblclick_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r88); const i_r69 = restoredCtx.$implicit; const index_r70 = restoredCtx.index; const ctx_r89 = ɵngcc0.ɵɵnextContext(); return ctx_r89._rowClick($event, i_r69, index_r70, true); });
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_template_9_td_1_Template, 1, 2, "td", 41);
    ɵngcc0.ɵɵtemplate(2, STComponent_ng_template_9_td_2_Template, 3, 11, "td", 42);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "tr", 43);
    ɵngcc0.ɵɵtemplate(4, STComponent_ng_template_9_ng_template_4_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r69 = ctx.$implicit;
    const index_r70 = ctx.index;
    const ctx_r8 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", i_r69._rowClassName);
    ɵngcc0.ɵɵattribute("data-index", index_r70);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r8.expand);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r8._columns);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("nzExpand", i_r69.expand);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r8.expand)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(7, _c5, i_r69, index_r70));
} }
function STComponent_ng_container_11_ng_container_1_ng_template_1_Template(rf, ctx) { }
function STComponent_ng_container_11_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_container_11_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r91 = ctx.$implicit;
    const index_r92 = ctx.index;
    ɵngcc0.ɵɵnextContext(2);
    const _r7 = ɵngcc0.ɵɵreference(10);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r7)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c5, i_r91, index_r92));
} }
function STComponent_ng_container_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_container_11_ng_container_1_Template, 2, 5, "ng-container", 10);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r9._data);
} }
function STComponent_ng_container_12_ng_template_1_ng_template_0_Template(rf, ctx) { }
function STComponent_ng_container_12_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, STComponent_ng_container_12_ng_template_1_ng_template_0_Template, 0, 0, "ng-template", 22);
} if (rf & 2) {
    const i_r95 = ctx.$implicit;
    const index_r96 = ctx.index;
    ɵngcc0.ɵɵnextContext(2);
    const _r7 = ɵngcc0.ɵɵreference(10);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r7)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c5, i_r95, index_r96));
} }
function STComponent_ng_container_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_container_12_ng_template_1_Template, 1, 5, "ng-template", 49);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function STComponent_ng_container_13_ng_template_1_Template(rf, ctx) { }
function STComponent_ng_container_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_container_13_ng_template_1_Template, 0, 0, "ng-template", 22);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r11.body)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c4, ctx_r11._statistical));
} }
function STComponent_ng_template_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    const range_r99 = ctx.range;
    const total_r100 = ctx.$implicit;
    const ctx_r13 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵtextInterpolate(ctx_r13.renderTotal(total_r100, range_r99));
} }
function STComponent_ng_container_19_li_1_Template(rf, ctx) { if (rf & 1) {
    const _r106 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 38);
    ɵngcc0.ɵɵlistener("click", function STComponent_ng_container_19_li_1_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r106); const i_r101 = ɵngcc0.ɵɵnextContext().$implicit; return i_r101.fn(i_r101); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r101 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", i_r101.text, ɵngcc0.ɵɵsanitizeHtml);
} }
function STComponent_ng_container_19_li_2_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r111 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 38);
    ɵngcc0.ɵɵlistener("click", function STComponent_ng_container_19_li_2_li_2_Template_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r111); const ci_r109 = restoredCtx.$implicit; return ci_r109.fn(ci_r109); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ci_r109 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", ci_r109.text, ɵngcc0.ɵɵsanitizeHtml);
} }
function STComponent_ng_container_19_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 52);
    ɵngcc0.ɵɵelementStart(1, "ul");
    ɵngcc0.ɵɵtemplate(2, STComponent_ng_container_19_li_2_li_2_Template, 1, 1, "li", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r101 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("nzTitle", i_r101.text);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", i_r101.children);
} }
function STComponent_ng_container_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STComponent_ng_container_19_li_1_Template, 1, 1, "li", 50);
    ɵngcc0.ɵɵtemplate(2, STComponent_ng_container_19_li_2_Template, 3, 2, "li", 51);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r101 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r101.children.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r101.children.length > 0);
} }
function STFilterComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "input", 13);
    ɵngcc0.ɵɵlistener("ngModelChange", function STFilterComponent_div_6_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.f.menus[0].value = $event; })("ngModelChange", function STFilterComponent_div_6_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.n.emit($event); })("keyup.enter", function STFilterComponent_div_6_Template_input_keyup_enter_1_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.confirm(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.f.menus[0].value);
    ɵngcc0.ɵɵattribute("placeholder", ctx_r1.f.placeholder);
} }
function STFilterComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵelementStart(1, "nz-input-number", 15);
    ɵngcc0.ɵɵlistener("ngModelChange", function STFilterComponent_div_7_Template_nz_input_number_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.f.menus[0].value = $event; })("ngModelChange", function STFilterComponent_div_7_Template_nz_input_number_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.n.emit($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r2.f.menus[0].value)("nzMin", ctx_r2.f.number.min)("nzMax", ctx_r2.f.number.max)("nzStep", ctx_r2.f.number.step)("nzPrecision", ctx_r2.f.number.precision)("nzPlaceHolder", ctx_r2.f.placeholder);
} }
function STFilterComponent_div_8_nz_date_picker_1_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "nz-date-picker", 18);
    ɵngcc0.ɵɵlistener("ngModelChange", function STFilterComponent_div_8_nz_date_picker_1_Template_nz_date_picker_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.f.menus[0].value = $event; })("ngModelChange", function STFilterComponent_div_8_nz_date_picker_1_Template_nz_date_picker_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.n.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("nzMode", ctx_r15.f.date.mode)("ngModel", ctx_r15.f.menus[0].value)("nzShowNow", ctx_r15.f.date.showNow)("nzShowToday", ctx_r15.f.date.showToday)("nzDisabledDate", ctx_r15.f.date.disabledDate)("nzDisabledTime", ctx_r15.f.date.disabledTime);
} }
function STFilterComponent_div_8_nz_range_picker_2_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "nz-range-picker", 18);
    ɵngcc0.ɵɵlistener("ngModelChange", function STFilterComponent_div_8_nz_range_picker_2_Template_nz_range_picker_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r20 = ɵngcc0.ɵɵnextContext(2); return ctx_r20.f.menus[0].value = $event; })("ngModelChange", function STFilterComponent_div_8_nz_range_picker_2_Template_nz_range_picker_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r22 = ɵngcc0.ɵɵnextContext(2); return ctx_r22.n.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("nzMode", ctx_r16.f.date.mode)("ngModel", ctx_r16.f.menus[0].value)("nzShowNow", ctx_r16.f.date.showNow)("nzShowToday", ctx_r16.f.date.showToday)("nzDisabledDate", ctx_r16.f.date.disabledDate)("nzDisabledTime", ctx_r16.f.date.disabledTime);
} }
function STFilterComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵtemplate(1, STFilterComponent_div_8_nz_date_picker_1_Template, 1, 6, "nz-date-picker", 17);
    ɵngcc0.ɵɵtemplate(2, STFilterComponent_div_8_nz_range_picker_2_Template, 1, 6, "nz-range-picker", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.f.date.range);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.f.date.range);
} }
function STFilterComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 19);
} }
function STFilterComponent_div_10_ng_template_1_Template(rf, ctx) { }
const _c6 = function (a0, a1) { return { $implicit: a0, col: a1 }; };
function STFilterComponent_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵtemplate(1, STFilterComponent_div_10_ng_template_1_Template, 0, 0, "ng-template", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r5.f.custom)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c6, ctx_r5.f, ctx_r5.col));
} }
function STFilterComponent_ul_11_ng_container_1_li_1_Template(rf, ctx) { if (rf & 1) {
    const _r29 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 25);
    ɵngcc0.ɵɵelementStart(1, "label", 26);
    ɵngcc0.ɵɵlistener("ngModelChange", function STFilterComponent_ul_11_ng_container_1_li_1_Template_label_ngModelChange_1_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r29); const filter_r27 = restoredCtx.$implicit; return filter_r27.checked = $event; })("ngModelChange", function STFilterComponent_ul_11_ng_container_1_li_1_Template_label_ngModelChange_1_listener() { ɵngcc0.ɵɵrestoreView(_r29); const ctx_r30 = ɵngcc0.ɵɵnextContext(3); return ctx_r30.checkboxChange(); });
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const filter_r27 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", filter_r27.checked);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", filter_r27.text, " ");
} }
function STFilterComponent_ul_11_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STFilterComponent_ul_11_ng_container_1_li_1_Template, 3, 2, "li", 24);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r24 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r24.f.menus);
} }
function STFilterComponent_ul_11_ng_container_2_li_1_Template(rf, ctx) { if (rf & 1) {
    const _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 25);
    ɵngcc0.ɵɵelementStart(1, "label", 27);
    ɵngcc0.ɵɵlistener("ngModelChange", function STFilterComponent_ul_11_ng_container_2_li_1_Template_label_ngModelChange_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r34); const filter_r32 = restoredCtx.$implicit; const ctx_r33 = ɵngcc0.ɵɵnextContext(3); return ctx_r33.radioChange(filter_r32); });
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const filter_r32 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", filter_r32.checked);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", filter_r32.text, " ");
} }
function STFilterComponent_ul_11_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STFilterComponent_ul_11_ng_container_2_li_1_Template, 3, 2, "li", 24);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r25 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r25.f.menus);
} }
function STFilterComponent_ul_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 22);
    ɵngcc0.ɵɵtemplate(1, STFilterComponent_ul_11_ng_container_1_Template, 2, 1, "ng-container", 23);
    ɵngcc0.ɵɵtemplate(2, STFilterComponent_ul_11_ng_container_2_Template, 2, 1, "ng-container", 23);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.f.multiple);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r6.f.multiple);
} }
function STFilterComponent_div_12_Template(rf, ctx) { if (rf & 1) {
    const _r36 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵelementStart(1, "a", 29);
    ɵngcc0.ɵɵlistener("click", function STFilterComponent_div_12_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r36); const ctx_r35 = ɵngcc0.ɵɵnextContext(); return ctx_r35.visible = false; });
    ɵngcc0.ɵɵelementStart(2, "span", 30);
    ɵngcc0.ɵɵlistener("click", function STFilterComponent_div_12_Template_span_click_2_listener() { ɵngcc0.ɵɵrestoreView(_r36); const ctx_r37 = ɵngcc0.ɵɵnextContext(); return ctx_r37.confirm(); });
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "a", 31);
    ɵngcc0.ɵɵlistener("click", function STFilterComponent_div_12_Template_a_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r36); const ctx_r38 = ɵngcc0.ɵɵnextContext(); return ctx_r38.visible = false; });
    ɵngcc0.ɵɵelementStart(5, "span", 30);
    ɵngcc0.ɵɵlistener("click", function STFilterComponent_div_12_Template_span_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r36); const ctx_r39 = ɵngcc0.ɵɵnextContext(); return ctx_r39.reset(); });
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r7.f.confirmText || ctx_r7.locale.filterConfirm);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r7.f.clearText || ctx_r7.locale.filterReset);
} }
function STTdComponent_ng_template_0_ng_container_0_ng_template_1_Template(rf, ctx) { }
function STTdComponent_ng_template_0_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_template_0_ng_container_0_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c4, i_r9));
} }
function STTdComponent_ng_template_0_span_1_ng_template_1_Template(rf, ctx) { }
function STTdComponent_ng_template_0_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 8);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_template_0_span_1_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵproperty("nzTooltipTitle", i_r9.tooltip);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c4, i_r9));
} }
function STTdComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, STTdComponent_ng_template_0_ng_container_0_Template, 2, 4, "ng-container", 6);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_template_0_span_1_Template, 2, 5, "span", 7);
} if (rf & 2) {
    const i_r9 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", !i_r9.tooltip);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r9.tooltip);
} }
function STTdComponent_ng_template_2_a_0_ng_template_1_Template(rf, ctx) { }
function STTdComponent_ng_template_2_a_0_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 11);
    ɵngcc0.ɵɵlistener("nzOnConfirm", function STTdComponent_ng_template_2_a_0_Template_a_nzOnConfirm_0_listener() { ɵngcc0.ɵɵrestoreView(_r22); const i_r16 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r20 = ɵngcc0.ɵɵnextContext(); return ctx_r20._btn(i_r16); })("click", function STTdComponent_ng_template_2_a_0_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r23 = ɵngcc0.ɵɵnextContext(2); return ctx_r23._stopPropagation($event); });
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_template_2_a_0_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r16 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext();
    const _r4 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵproperty("nzPopconfirmTitle", i_r16.pop.title)("nzIcon", i_r16.pop.icon)("nzCondition", i_r16.pop.condition(i_r16))("nzCancelText", i_r16.pop.cancelText)("nzOkText", i_r16.pop.okText)("nzOkType", i_r16.pop.okType)("ngClass", i_r16.className);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r4)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(9, _c4, i_r16));
} }
function STTdComponent_ng_template_2_a_1_ng_template_1_Template(rf, ctx) { }
function STTdComponent_ng_template_2_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r28 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 12);
    ɵngcc0.ɵɵlistener("click", function STTdComponent_ng_template_2_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r28); const i_r16 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r26 = ɵngcc0.ɵɵnextContext(); return ctx_r26._btn(i_r16, $event); });
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_template_2_a_1_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r16 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext();
    const _r4 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵproperty("ngClass", i_r16.className);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r4)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c4, i_r16));
} }
function STTdComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, STTdComponent_ng_template_2_a_0_Template, 2, 11, "a", 9);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_template_2_a_1_Template, 2, 5, "a", 10);
} if (rf & 2) {
    const i_r16 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", i_r16.pop);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !i_r16.pop);
} }
function STTdComponent_ng_template_4_ng_container_0_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 16);
} if (rf & 2) {
    const i_r30 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("nzType", i_r30.icon.type)("nzTheme", i_r30.icon.theme)("nzSpin", i_r30.icon.spin)("nzTwotoneColor", i_r30.icon.twoToneColor);
} }
function STTdComponent_ng_template_4_ng_container_0_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 17);
} if (rf & 2) {
    const i_r30 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("nzIconfont", i_r30.icon.iconfont);
} }
function STTdComponent_ng_template_4_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_template_4_ng_container_0_i_1_Template, 1, 4, "i", 14);
    ɵngcc0.ɵɵtemplate(2, STTdComponent_ng_template_4_ng_container_0_i_2_Template, 1, 1, "i", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r30 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !i_r30.icon.iconfont);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r30.icon.iconfont);
} }
const _c7 = function (a0) { return { "pl-xs": a0 }; };
function STTdComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, STTdComponent_ng_template_4_ng_container_0_Template, 3, 2, "ng-container", 6);
    ɵngcc0.ɵɵelement(1, "span", 13);
} if (rf & 2) {
    const i_r30 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", i_r30.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", i_r30._text, ɵngcc0.ɵɵsanitizeHtml)("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c7, i_r30.icon));
} }
function STTdComponent_ng_template_6_Template(rf, ctx) { }
function STTdComponent_ng_container_8_label_2_Template(rf, ctx) { if (rf & 1) {
    const _r45 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 25);
    ɵngcc0.ɵɵlistener("ngModelChange", function STTdComponent_ng_container_8_label_2_Template_label_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r45); const ctx_r44 = ɵngcc0.ɵɵnextContext(2); return ctx_r44._checkbox($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r37 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("nzDisabled", ctx_r37.i.disabled)("ngModel", ctx_r37.i.checked);
} }
function STTdComponent_ng_container_8_label_3_Template(rf, ctx) { if (rf & 1) {
    const _r47 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 26);
    ɵngcc0.ɵɵlistener("ngModelChange", function STTdComponent_ng_container_8_label_3_Template_label_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r47); const ctx_r46 = ɵngcc0.ɵɵnextContext(2); return ctx_r46._radio($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r38 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("nzDisabled", ctx_r38.i.disabled)("ngModel", ctx_r38.i.checked);
} }
function STTdComponent_ng_container_8_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r49 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 27);
    ɵngcc0.ɵɵlistener("click", function STTdComponent_ng_container_8_a_4_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r49); const ctx_r48 = ɵngcc0.ɵɵnextContext(2); return ctx_r48._link($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r39 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("innerHTML", ctx_r39.i._values[ctx_r39.cIdx]._text, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵattribute("title", ctx_r39.i._values[ctx_r39.cIdx].text);
} }
function STTdComponent_ng_container_8_ng_container_5_nz_tag_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "nz-tag", 30);
    ɵngcc0.ɵɵelement(1, "span", 31);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r50 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("nzColor", ctx_r50.i._values[ctx_r50.cIdx].color);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", ctx_r50.i._values[ctx_r50.cIdx]._text, ɵngcc0.ɵɵsanitizeHtml);
} }
function STTdComponent_ng_container_8_ng_container_5_nz_badge_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nz-badge", 32);
} if (rf & 2) {
    const ctx_r51 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("nzStatus", ctx_r51.i._values[ctx_r51.cIdx].color)("nzText", ctx_r51.i._values[ctx_r51.cIdx].text);
} }
function STTdComponent_ng_container_8_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_container_8_ng_container_5_nz_tag_1_Template, 2, 2, "nz-tag", 28);
    ɵngcc0.ɵɵtemplate(2, STTdComponent_ng_container_8_ng_container_5_nz_badge_2_Template, 1, 2, "nz-badge", 29);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "tag");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "badge");
} }
function STTdComponent_ng_container_8_6_ng_template_0_Template(rf, ctx) { }
function STTdComponent_ng_container_8_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, STTdComponent_ng_container_8_6_ng_template_0_Template, 0, 0, "ng-template", 33);
} if (rf & 2) {
    const ctx_r41 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("record", ctx_r41.i)("column", ctx_r41.c);
} }
function STTdComponent_ng_container_8_ng_container_7_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 31);
} if (rf & 2) {
    const ctx_r53 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("innerHTML", ctx_r53.i._values[ctx_r53.cIdx]._text, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵattribute("title", ctx_r53.c._isTruncate ? ctx_r53.i._values[ctx_r53.cIdx].text : null);
} }
function STTdComponent_ng_container_8_ng_container_7_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 36);
} if (rf & 2) {
    const ctx_r54 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("innerText", ctx_r54.i._values[ctx_r54.cIdx]._text);
    ɵngcc0.ɵɵattribute("title", ctx_r54.c._isTruncate ? ctx_r54.i._values[ctx_r54.cIdx].text : null);
} }
function STTdComponent_ng_container_8_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_container_8_ng_container_7_span_1_Template, 1, 2, "span", 34);
    ɵngcc0.ɵɵtemplate(2, STTdComponent_ng_container_8_ng_container_7_span_2_Template, 1, 2, "span", 35);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r42 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r42.c.safeType !== "text");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r42.c.safeType === "text");
} }
function STTdComponent_ng_container_8_ng_container_8_a_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a", 42);
    ɵngcc0.ɵɵelement(1, "span", 31);
    ɵngcc0.ɵɵelement(2, "i", 43);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const btn_r55 = ɵngcc0.ɵɵnextContext().$implicit;
    const _r58 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵproperty("nzDropdownMenu", _r58);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", btn_r55._text, ɵngcc0.ɵɵsanitizeHtml);
} }
function STTdComponent_ng_container_8_ng_container_8_ng_container_5_li_1_ng_template_1_Template(rf, ctx) { }
function STTdComponent_ng_container_8_ng_container_8_ng_container_5_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 46);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_container_8_ng_container_8_ng_container_5_li_1_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const subBtn_r63 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext(3);
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵclassProp("st__btn-disabled", subBtn_r63._disabled);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(4, _c4, subBtn_r63));
} }
function STTdComponent_ng_container_8_ng_container_8_ng_container_5_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 47);
} }
function STTdComponent_ng_container_8_ng_container_8_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_container_8_ng_container_8_ng_container_5_li_1_Template, 2, 6, "li", 44);
    ɵngcc0.ɵɵtemplate(2, STTdComponent_ng_container_8_ng_container_8_ng_container_5_li_2_Template, 1, 0, "li", 45);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const subBtn_r63 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", subBtn_r63.type !== "divider");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", subBtn_r63.type === "divider");
} }
function STTdComponent_ng_container_8_ng_container_8_span_6_ng_template_1_Template(rf, ctx) { }
function STTdComponent_ng_container_8_ng_container_8_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_container_8_ng_container_8_span_6_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const btn_r55 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext(2);
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵclassProp("st__btn-disabled", btn_r55._disabled);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(4, _c4, btn_r55));
} }
function STTdComponent_ng_container_8_ng_container_8_nz_divider_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nz-divider", 48);
} }
function STTdComponent_ng_container_8_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, STTdComponent_ng_container_8_ng_container_8_a_1_Template, 3, 2, "a", 37);
    ɵngcc0.ɵɵelementStart(2, "nz-dropdown-menu", null, 38);
    ɵngcc0.ɵɵelementStart(4, "ul", 39);
    ɵngcc0.ɵɵtemplate(5, STTdComponent_ng_container_8_ng_container_8_ng_container_5_Template, 3, 2, "ng-container", 24);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, STTdComponent_ng_container_8_ng_container_8_span_6_Template, 2, 6, "span", 40);
    ɵngcc0.ɵɵtemplate(7, STTdComponent_ng_container_8_ng_container_8_nz_divider_7_Template, 1, 0, "nz-divider", 41);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const btn_r55 = ctx.$implicit;
    const last_r56 = ctx.last;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", btn_r55.children.length > 0);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngForOf", btn_r55.children);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", btn_r55.children.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !last_r56);
} }
function STTdComponent_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementContainerStart(1, 18);
    ɵngcc0.ɵɵtemplate(2, STTdComponent_ng_container_8_label_2_Template, 1, 2, "label", 19);
    ɵngcc0.ɵɵtemplate(3, STTdComponent_ng_container_8_label_3_Template, 1, 2, "label", 20);
    ɵngcc0.ɵɵtemplate(4, STTdComponent_ng_container_8_a_4_Template, 1, 2, "a", 21);
    ɵngcc0.ɵɵtemplate(5, STTdComponent_ng_container_8_ng_container_5_Template, 3, 2, "ng-container", 6);
    ɵngcc0.ɵɵtemplate(6, STTdComponent_ng_container_8_6_Template, 1, 2, undefined, 22);
    ɵngcc0.ɵɵtemplate(7, STTdComponent_ng_container_8_ng_container_7_Template, 3, 2, "ng-container", 23);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵtemplate(8, STTdComponent_ng_container_8_ng_container_8_Template, 8, 4, "ng-container", 24);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitch", ctx_r8.c.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "checkbox");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "radio");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "link");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r8.i._values[ctx_r8.cIdx].text);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "widget");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r8.i._values[ctx_r8.cIdx].buttons);
} }
const _c8 = function (a0, a1, a2) { return { $implicit: a0, index: a1, column: a2 }; };
class STRowSource {
    constructor() {
        this.titles = {};
        this.rows = {};
    }
    add(type, path, ref) {
        this[type === 'title' ? 'titles' : 'rows'][path] = ref;
    }
    getTitle(path) {
        return this.titles[path];
    }
    getRow(path) {
        return this.rows[path];
    }
}
STRowSource.ɵfac = function STRowSource_Factory(t) { return new (t || STRowSource)(); };
STRowSource.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: STRowSource, factory: STRowSource.ɵfac });
class STRowDirective {
    constructor(ref, source) {
        this.ref = ref;
        this.source = source;
    }
    ngOnInit() {
        this.source.add(this.type, this.id, this.ref);
    }
}
STRowDirective.ɵfac = function STRowDirective_Factory(t) { return new (t || STRowDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(STRowSource, 1)); };
STRowDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: STRowDirective, selectors: [["", "st-row", ""]], inputs: { id: ["st-row", "id"], type: "type" } });
STRowDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: STRowSource, decorators: [{ type: Host }] }
];
STRowDirective.propDecorators = {
    id: [{ type: Input, args: ['st-row',] }],
    type: [{ type: Input }]
};

class STWidgetRegistry {
    constructor() {
        this._widgets = {};
    }
    get widgets() {
        return this._widgets;
    }
    register(type, widget) {
        this._widgets[type] = widget;
    }
    has(type) {
        return this._widgets.hasOwnProperty(type);
    }
    get(type) {
        return this._widgets[type];
    }
}
STWidgetRegistry.ɵfac = function STWidgetRegistry_Factory(t) { return new (t || STWidgetRegistry)(); };
STWidgetRegistry.ɵprov = i0.ɵɵdefineInjectable({ factory: function STWidgetRegistry_Factory() { return new STWidgetRegistry(); }, token: STWidgetRegistry, providedIn: "root" });

class STColumnSource {
    constructor(dom, rowSource, acl, i18nSrv, stWidgetRegistry) {
        this.dom = dom;
        this.rowSource = rowSource;
        this.acl = acl;
        this.i18nSrv = i18nSrv;
        this.stWidgetRegistry = stWidgetRegistry;
    }
    setCog(val) {
        this.cog = val;
    }
    fixPop(i, def) {
        if (i.pop == null || i.pop === false) {
            i.pop = false;
            return;
        }
        let pop = Object.assign({}, def);
        if (typeof i.pop === 'string') {
            pop.title = i.pop;
        }
        else if (typeof i.pop === 'object') {
            pop = Object.assign(Object.assign({}, pop), i.pop);
        }
        if (typeof pop.condition !== 'function') {
            pop.condition = () => false;
        }
        i.pop = pop;
    }
    btnCoerce(list) {
        if (!list)
            return [];
        const ret = [];
        const { modal, drawer, pop, btnIcon } = this.cog;
        for (const item of list) {
            if (this.acl && item.acl && !this.acl.can(item.acl)) {
                continue;
            }
            if (item.type === 'modal' || item.type === 'static') {
                if (item.modal == null || item.modal.component == null) {
                    console.warn(`[st] Should specify modal parameter when type is modal or static`);
                    item.type = 'none';
                }
                else {
                    item.modal = Object.assign(Object.assign({ paramsName: 'record', size: 'lg' }, modal), item.modal);
                }
            }
            if (item.type === 'drawer') {
                if (item.drawer == null || item.drawer.component == null) {
                    console.warn(`[st] Should specify drawer parameter when type is drawer`);
                    item.type = 'none';
                }
                else {
                    item.drawer = Object.assign(Object.assign({ paramsName: 'record', size: 'lg' }, drawer), item.drawer);
                }
            }
            if (item.type === 'del' && typeof item.pop === 'undefined') {
                item.pop = true;
            }
            // pop
            this.fixPop(item, pop);
            if (item.icon) {
                item.icon = Object.assign(Object.assign({}, btnIcon), (typeof item.icon === 'string' ? { type: item.icon } : item.icon));
            }
            item.children = item.children && item.children.length > 0 ? this.btnCoerce(item.children) : [];
            // i18n
            if (item.i18n && this.i18nSrv) {
                item.text = this.i18nSrv.fanyi(item.i18n);
            }
            ret.push(item);
        }
        this.btnCoerceIf(ret);
        return ret;
    }
    btnCoerceIf(list) {
        for (const item of list) {
            if (!item.iif)
                item.iif = () => true;
            item.iifBehavior = item.iifBehavior || this.cog.iifBehavior;
            if (item.children && item.children.length > 0) {
                this.btnCoerceIf(item.children);
            }
            else {
                item.children = [];
            }
        }
    }
    fixedCoerce(list) {
        const countReduce = (a, b) => a + +b.width.toString().replace('px', '');
        // left width
        list
            .filter(w => w.fixed && w.fixed === 'left' && w.width)
            .forEach((item, idx) => (item._left = `${list.slice(0, idx).reduce(countReduce, 0)}px`));
        // right width
        list
            .filter(w => w.fixed && w.fixed === 'right' && w.width)
            .reverse()
            .forEach((item, idx) => (item._right = `${idx > 0 ? list.slice(-idx).reduce(countReduce, 0) : 0}px`));
    }
    sortCoerce(item) {
        const res = this.fixSortCoerce(item);
        res.reName = Object.assign(Object.assign({}, this.cog.sortReName), res.reName);
        return res;
    }
    fixSortCoerce(item) {
        if (typeof item.sort === 'undefined') {
            return { enabled: false };
        }
        let res = {};
        if (typeof item.sort === 'string') {
            res.key = item.sort;
        }
        else if (typeof item.sort !== 'boolean') {
            res = item.sort;
        }
        else if (typeof item.sort === 'boolean') {
            res.compare = (a, b) => a[item.indexKey] - b[item.indexKey];
        }
        if (!res.key) {
            res.key = item.indexKey;
        }
        res.enabled = true;
        return res;
    }
    filterCoerce(item) {
        if (item.filter == null) {
            return null;
        }
        let res = item.filter;
        res.type = res.type || 'default';
        res.showOPArea = res.showOPArea !== false;
        let icon = 'filter';
        let iconTheme = 'fill';
        let fixMenus = true;
        let value = undefined;
        switch (res.type) {
            case 'keyword':
                icon = 'search';
                iconTheme = 'outline';
                break;
            case 'number':
                icon = 'search';
                iconTheme = 'outline';
                res.number = Object.assign({ step: 1, min: -Infinity, max: Infinity }, res.number);
                break;
            case 'date':
                icon = 'calendar';
                iconTheme = 'outline';
                res.date = Object.assign({ range: false, mode: 'date', showToday: true, showNow: false }, res.date);
                break;
            default:
                fixMenus = false;
                break;
        }
        if (fixMenus && (res.menus == null || res.menus.length === 0)) {
            res.menus = [{ value }];
        }
        if (res.menus.length === 0) {
            return null;
        }
        if (typeof res.multiple === 'undefined') {
            res.multiple = true;
        }
        res.confirmText = res.confirmText || this.cog.filterConfirmText;
        res.clearText = res.clearText || this.cog.filterClearText;
        res.key = res.key || item.indexKey;
        res.icon = res.icon || icon;
        const baseIcon = { type: icon, theme: iconTheme };
        if (typeof res.icon === 'string') {
            res.icon = Object.assign(Object.assign({}, baseIcon), { type: res.icon });
        }
        else {
            res.icon = Object.assign(Object.assign({}, baseIcon), res.icon);
        }
        this.updateDefault(res);
        if (this.acl) {
            res.menus = res.menus.filter(w => this.acl.can(w.acl));
        }
        if (res.menus.length <= 0) {
            res = null;
        }
        return res;
    }
    restoreRender(item) {
        if (item.renderTitle) {
            item.__renderTitle =
                typeof item.renderTitle === 'string'
                    ? this.rowSource.getTitle(item.renderTitle)
                    : item.renderTitle;
        }
        if (item.render) {
            item.__render =
                typeof item.render === 'string' ? this.rowSource.getRow(item.render) : item.render;
        }
    }
    widgetCoerce(item) {
        var _a;
        if (item.type !== 'widget')
            return;
        if (item.widget == null || !this.stWidgetRegistry.has(item.widget.type)) {
            delete item.type;
            warn(`st: No widget for type "${(_a = item.widget) === null || _a === void 0 ? void 0 : _a.type}"`);
        }
    }
    genHeaders(rootColumns) {
        const rows = [];
        const widths = [];
        const fillRowCells = (columns, colIndex, rowIndex = 0) => {
            // Init rows
            rows[rowIndex] = rows[rowIndex] || [];
            let currentColIndex = colIndex;
            const colSpans = columns.map(column => {
                const cell = {
                    column,
                    colStart: currentColIndex,
                    hasSubColumns: false
                };
                let colSpan = 1;
                const subColumns = column.children;
                if (Array.isArray(subColumns) && subColumns.length > 0) {
                    colSpan = fillRowCells(subColumns, currentColIndex, rowIndex + 1).reduce((total, count) => total + count, 0);
                    cell.hasSubColumns = true;
                }
                else {
                    widths.push(cell.column.width || '');
                }
                if ('colSpan' in column) {
                    colSpan = column.colSpan;
                }
                if ('rowSpan' in column) {
                    cell.rowSpan = column.rowSpan;
                }
                cell.colSpan = colSpan;
                cell.colEnd = cell.colStart + colSpan - 1;
                rows[rowIndex].push(cell);
                currentColIndex += colSpan;
                return colSpan;
            });
            return colSpans;
        };
        fillRowCells(rootColumns, 0);
        // Handle `rowSpan`
        const rowCount = rows.length;
        for (let rowIndex = 0; rowIndex < rowCount; rowIndex += 1) {
            rows[rowIndex].forEach(cell => {
                if (!('rowSpan' in cell) && !cell.hasSubColumns) {
                    cell.rowSpan = rowCount - rowIndex;
                }
            });
        }
        return { headers: rows, headerWidths: rowCount > 1 ? widths : null };
    }
    cleanCond(list) {
        const res = [];
        const copyList = deepCopy(list);
        for (const item of copyList) {
            if (item.iif && !item.iif(item)) {
                continue;
            }
            if (this.acl && item.acl && !this.acl.can(item.acl)) {
                continue;
            }
            if (Array.isArray(item.children) && item.children.length > 0) {
                item.children = this.cleanCond(item.children);
            }
            res.push(item);
        }
        return res;
    }
    process(list, options) {
        if (!list || list.length === 0)
            throw new Error(`[st]: the columns property muse be define!`);
        const { noIndex } = this.cog;
        let checkboxCount = 0;
        let radioCount = 0;
        let point = 0;
        const columns = [];
        const processItem = (item) => {
            var _a, _b;
            // index
            if (item.index) {
                if (!Array.isArray(item.index)) {
                    item.index = item.index.toString().split('.');
                }
                item.indexKey = item.index.join('.');
            }
            // #region title
            const tit = (typeof item.title === 'string' ? { text: item.title } : item.title) || {};
            if (tit.i18n && this.i18nSrv) {
                tit.text = this.i18nSrv.fanyi(tit.i18n);
            }
            if (tit.text) {
                tit._text = this.dom.bypassSecurityTrustHtml(tit.text);
            }
            item.title = tit;
            // #endregion
            // no
            if (item.type === 'no') {
                item.noIndex = item.noIndex == null ? noIndex : item.noIndex;
            }
            // checkbox
            if (item.selections == null) {
                item.selections = [];
            }
            if (item.type === 'checkbox') {
                ++checkboxCount;
                if (!item.width) {
                    item.width = `${item.selections.length > 0 ? 62 : 50}px`;
                }
            }
            if (this.acl) {
                item.selections = item.selections.filter(w => this.acl.can(w.acl));
            }
            // radio
            if (item.type === 'radio') {
                ++radioCount;
                item.selections = [];
                if (!item.width) {
                    item.width = '50px';
                }
            }
            // types
            if (item.type === 'yn') {
                item.yn = Object.assign(Object.assign({ truth: true }, this.cog.yn), item.yn);
            }
            // date
            if (item.type === 'date') {
                item.dateFormat = item.dateFormat || ((_a = this.cog.date) === null || _a === void 0 ? void 0 : _a.format);
            }
            if ((item.type === 'link' && typeof item.click !== 'function') ||
                (item.type === 'badge' && item.badge == null) ||
                (item.type === 'tag' && item.tag == null) ||
                (item.type === 'enum' && item.enum == null)) {
                item.type = '';
            }
            item._isTruncate = !!item.width && options.widthMode.strictBehavior === 'truncate' && item.type !== 'img';
            // className
            if (!item.className) {
                item.className = {
                    number: 'text-right',
                    currency: 'text-right',
                    date: 'text-center'
                }[item.type];
            }
            item._className = item.className || (item._isTruncate ? 'text-truncate' : null);
            // width
            if (typeof item.width === 'number') {
                item._width = item.width;
                item.width = `${item.width}px`;
            }
            item._left = false;
            item._right = false;
            item.safeType = (_b = item.safeType) !== null && _b !== void 0 ? _b : options.safeType;
            // sorter
            item._sort = this.sortCoerce(item);
            // filter
            item.filter = this.filterCoerce(item);
            // buttons
            item.buttons = this.btnCoerce(item.buttons);
            // widget
            this.widgetCoerce(item);
            // restore custom row
            this.restoreRender(item);
            // resizable
            item.resizable = Object.assign(Object.assign({ disabled: true, bounds: 'window', minWidth: 60, maxWidth: 360, preview: true }, options.resizable), (typeof item.resizable === 'boolean' ? { disabled: !item.resizable } : item.resizable));
            item.__point = point++;
            return item;
        };
        const processList = (data) => {
            for (const item of data) {
                columns.push(processItem(item));
                if (Array.isArray(item.children)) {
                    processList(item.children);
                }
            }
        };
        const copyList = this.cleanCond(list);
        processList(copyList);
        if (checkboxCount > 1) {
            throw new Error(`[st]: just only one column checkbox`);
        }
        if (radioCount > 1) {
            throw new Error(`[st]: just only one column radio`);
        }
        this.fixedCoerce(columns);
        return Object.assign({ columns: columns.filter(w => !Array.isArray(w.children) || w.children.length === 0) }, this.genHeaders(copyList));
    }
    restoreAllRender(columns) {
        columns.forEach(i => this.restoreRender(i));
    }
    updateDefault(filter) {
        if (filter.type === 'default') {
            filter.default = filter.menus.findIndex(w => w.checked) !== -1;
        }
        else {
            filter.default = !!filter.menus[0].value;
        }
        return this;
    }
    cleanFilter(col) {
        const f = col.filter;
        f.default = false;
        if (f.type === 'default') {
            f.menus.forEach(i => (i.checked = false));
        }
        else {
            f.menus[0].value = undefined;
        }
        return this;
    }
}
STColumnSource.ɵfac = function STColumnSource_Factory(t) { return new (t || STColumnSource)(ɵngcc0.ɵɵinject(ɵngcc1.DomSanitizer), ɵngcc0.ɵɵinject(STRowSource, 1), ɵngcc0.ɵɵinject(ɵngcc2.ACLService, 8), ɵngcc0.ɵɵinject(ALAIN_I18N_TOKEN, 8), ɵngcc0.ɵɵinject(STWidgetRegistry)); };
STColumnSource.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: STColumnSource, factory: STColumnSource.ɵfac });
STColumnSource.ctorParameters = () => [
    { type: DomSanitizer },
    { type: STRowSource, decorators: [{ type: Host }] },
    { type: ACLService, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [ALAIN_I18N_TOKEN,] }] },
    { type: STWidgetRegistry }
];

class STDataSource {
    constructor(http, datePipe, ynPipe, numberPipe, currencySrv, dom) {
        this.http = http;
        this.datePipe = datePipe;
        this.ynPipe = ynPipe;
        this.numberPipe = numberPipe;
        this.currencySrv = currencySrv;
        this.dom = dom;
        this.sortTick = 0;
    }
    process(options) {
        let data$;
        let isRemote = false;
        const { data, res, total, page, pi, ps, paginator, columns } = options;
        let retTotal;
        let retPs;
        let retList;
        let retPi;
        let rawData;
        let showPage = page.show;
        if (typeof data === 'string') {
            isRemote = true;
            data$ = this.getByRemote(data, options).pipe(map(result => {
                rawData = result;
                let ret;
                if (Array.isArray(result)) {
                    ret = result;
                    retTotal = ret.length;
                    retPs = retTotal;
                    showPage = false;
                }
                else {
                    // list
                    ret = deepGet(result, res.reName.list, []);
                    if (ret == null || !Array.isArray(ret)) {
                        ret = [];
                    }
                    // total
                    const resultTotal = res.reName.total && deepGet(result, res.reName.total, null);
                    retTotal = resultTotal == null ? total || 0 : +resultTotal;
                }
                return deepCopy(ret);
            }));
        }
        else if (Array.isArray(data)) {
            data$ = of(data);
        }
        else {
            // a cold observable
            data$ = data;
        }
        if (!isRemote) {
            data$ = data$.pipe(
            // sort
            map((result) => {
                rawData = result;
                let copyResult = deepCopy(result);
                const sorterFn = this.getSorterFn(columns);
                if (sorterFn) {
                    copyResult = copyResult.sort(sorterFn);
                }
                return copyResult;
            }), 
            // filter
            map((result) => {
                columns
                    .filter(w => w.filter)
                    .forEach(c => {
                    const filter = c.filter;
                    const values = this.getFilteredData(filter);
                    if (values.length === 0)
                        return;
                    const onFilter = filter.fn;
                    if (typeof onFilter !== 'function') {
                        console.warn(`[st] Muse provide the fn function in filter`);
                        return;
                    }
                    result = result.filter(record => values.some(v => onFilter(v, record)));
                });
                return result;
            }), 
            // paging
            map((result) => {
                if (paginator && page.front) {
                    const maxPageIndex = Math.ceil(result.length / ps);
                    retPi = Math.max(1, pi > maxPageIndex ? maxPageIndex : pi);
                    retTotal = result.length;
                    if (page.show === true) {
                        return result.slice((retPi - 1) * ps, retPi * ps);
                    }
                }
                return result;
            }));
        }
        // pre-process
        if (typeof res.process === 'function') {
            data$ = data$.pipe(map(result => res.process(result, rawData)));
        }
        data$ = data$.pipe(map(result => this.optimizeData({ result, columns, rowClassName: options.rowClassName })));
        return data$.pipe(map(result => {
            retList = result;
            const realTotal = retTotal || total;
            const realPs = retPs || ps;
            return {
                pi: retPi,
                ps: retPs,
                total: retTotal,
                list: retList,
                statistical: this.genStatistical(columns, retList, rawData),
                pageShow: typeof showPage === 'undefined' ? realTotal > realPs : showPage
            };
        }));
    }
    get(item, col, idx) {
        var _a;
        try {
            const safeHtml = col.safeType === 'safeHtml';
            if (col.format) {
                const formatRes = col.format(item, col, idx) || '';
                return {
                    text: formatRes,
                    _text: safeHtml ? this.dom.bypassSecurityTrustHtml(formatRes) : formatRes,
                    org: formatRes,
                    safeType: col.safeType
                };
            }
            const value = deepGet(item, col.index, col.default);
            let text = value;
            let color;
            switch (col.type) {
                case 'no':
                    text = this.getNoIndex(item, col, idx);
                    break;
                case 'img':
                    text = value ? `<img src="${value}" class="img">` : '';
                    break;
                case 'number':
                    text = this.numberPipe.transform(value, col.numberDigits);
                    break;
                case 'currency':
                    text = this.currencySrv.format(value, (_a = col.currency) === null || _a === void 0 ? void 0 : _a.format);
                    break;
                case 'date':
                    text = value === col.default ? col.default : this.datePipe.transform(value, col.dateFormat);
                    break;
                case 'yn':
                    text = this.ynPipe.transform(value === col.yn.truth, col.yn.yes, col.yn.no, col.yn.mode, false);
                    break;
                case 'enum':
                    text = col.enum[value];
                    break;
                case 'tag':
                case 'badge':
                    const data = col.type === 'tag' ? col.tag : col.badge;
                    if (data && data[text]) {
                        const dataItem = data[text];
                        text = dataItem.text;
                        color = dataItem.color;
                    }
                    else {
                        text = '';
                    }
                    break;
            }
            if (text == null)
                text = '';
            return {
                text,
                _text: safeHtml ? this.dom.bypassSecurityTrustHtml(text) : text,
                org: value,
                color,
                safeType: col.safeType,
                buttons: []
            };
        }
        catch (ex) {
            const text = `INVALID DATA`;
            console.error(`Failed to get data`, item, col, ex);
            return { text, _text: text, org: text, buttons: [], safeType: 'text' };
        }
    }
    getByRemote(url, options) {
        const { req, page, paginator, pi, ps, singleSort, multiSort, columns } = options;
        const method = (req.method || 'GET').toUpperCase();
        let params = {};
        const reName = req.reName;
        if (paginator) {
            if (req.type === 'page') {
                params = {
                    [reName.pi]: page.zeroIndexed ? pi - 1 : pi,
                    [reName.ps]: ps
                };
            }
            else {
                params = {
                    [reName.skip]: (pi - 1) * ps,
                    [reName.limit]: ps
                };
            }
        }
        params = Object.assign(Object.assign(Object.assign(Object.assign({}, params), req.params), this.getReqSortMap(singleSort, multiSort, columns)), this.getReqFilterMap(columns));
        let reqOptions = {
            params,
            body: req.body,
            headers: req.headers
        };
        if (method === 'POST' && req.allInBody === true) {
            reqOptions = {
                body: Object.assign(Object.assign({}, req.body), params),
                headers: req.headers
            };
        }
        if (typeof req.process === 'function') {
            reqOptions = req.process(reqOptions);
        }
        if (!(reqOptions.params instanceof HttpParams)) {
            reqOptions.params = new HttpParams({ fromObject: reqOptions.params });
        }
        if (typeof options.customRequest === 'function') {
            return options.customRequest({ method, url, options: reqOptions });
        }
        return this.http.request(method, url, reqOptions);
    }
    optimizeData(options) {
        const { result, columns, rowClassName } = options;
        for (let i = 0, len = result.length; i < len; i++) {
            result[i]._values = columns.map(c => {
                if (Array.isArray(c.buttons) && c.buttons.length > 0) {
                    return { buttons: this.genButtons(c.buttons, result[i], c), _text: '' };
                }
                return this.get(result[i], c, i);
            });
            result[i]._rowClassName = [rowClassName ? rowClassName(result[i], i) : null, result[i].className]
                .filter(w => !!w)
                .join(' ');
        }
        return result;
    }
    getNoIndex(item, col, idx) {
        return typeof col.noIndex === 'function' ? col.noIndex(item, col, idx) : col.noIndex + idx;
    }
    genButtons(_btns, item, col) {
        const fn = (btns) => {
            return deepCopy(btns).filter(btn => {
                const result = btn.iif(item, btn, col);
                const isRenderDisabled = btn.iifBehavior === 'disabled';
                btn._result = result;
                btn._disabled = !result && isRenderDisabled;
                if (btn.children.length > 0) {
                    btn.children = fn(btn.children);
                }
                delete btn.iif;
                return result || isRenderDisabled;
            });
        };
        const res = fn(_btns);
        const fnText = (btns) => {
            for (const btn of btns) {
                btn._text = typeof btn.text === 'function' ? btn.text(item, btn) : btn.text || '';
                if (btn.children.length > 0) {
                    btn.children = fnText(btn.children);
                }
            }
            return btns;
        };
        return fnText(res);
    }
    // #region sort
    getValidSort(columns) {
        return columns.filter(item => item._sort && item._sort.enabled && item._sort.default).map(item => item._sort);
    }
    getSorterFn(columns) {
        const sortList = this.getValidSort(columns);
        if (sortList.length === 0) {
            return;
        }
        const sortItem = sortList[0];
        if (sortItem.compare === null) {
            return;
        }
        if (typeof sortItem.compare !== 'function') {
            console.warn(`[st] Muse provide the compare function in sort`);
            return;
        }
        return (a, b) => {
            const result = sortItem.compare(a, b);
            if (result !== 0) {
                return sortItem.default === 'descend' ? -result : result;
            }
            return 0;
        };
    }
    get nextSortTick() {
        return ++this.sortTick;
    }
    getReqSortMap(singleSort, multiSort, columns) {
        let ret = {};
        const sortList = this.getValidSort(columns);
        if (multiSort) {
            const ms = Object.assign({ key: 'sort', separator: '-', nameSeparator: '.', keepEmptyKey: true, arrayParam: false }, multiSort);
            const sortMap = sortList
                .sort((a, b) => a.tick - b.tick)
                .map(item => item.key + ms.nameSeparator + ((item.reName || {})[item.default] || item.default));
            ret = { [ms.key]: ms.arrayParam ? sortMap : sortMap.join(ms.separator) };
            return sortMap.length === 0 && ms.keepEmptyKey === false ? {} : ret;
        }
        if (sortList.length === 0)
            return ret;
        const mapData = sortList[0];
        let sortFiled = mapData.key;
        let sortValue = (sortList[0].reName || {})[mapData.default] || mapData.default;
        if (singleSort) {
            sortValue = sortFiled + (singleSort.nameSeparator || '.') + sortValue;
            sortFiled = singleSort.key || 'sort';
        }
        ret[sortFiled] = sortValue;
        return ret;
    }
    // #endregion
    // #region filter
    getFilteredData(filter) {
        return filter.type === 'default' ? filter.menus.filter(f => f.checked === true) : filter.menus.slice(0, 1);
    }
    getReqFilterMap(columns) {
        let ret = {};
        columns
            .filter(w => w.filter && w.filter.default === true)
            .forEach(col => {
            const filter = col.filter;
            const values = this.getFilteredData(filter);
            let obj = {};
            if (filter.reName) {
                obj = filter.reName(filter.menus, col);
            }
            else {
                obj[filter.key] = values.map(i => i.value).join(',');
            }
            ret = Object.assign(Object.assign({}, ret), obj);
        });
        return ret;
    }
    // #endregion
    // #region statistical
    genStatistical(columns, list, rawData) {
        const res = {};
        columns.forEach((col, index) => {
            res[col.key || col.indexKey || index] =
                col.statistical == null ? {} : this.getStatistical(col, index, list, rawData);
        });
        return res;
    }
    getStatistical(col, index, list, rawData) {
        var _a;
        const val = col.statistical;
        const item = Object.assign({ digits: 2, currency: undefined }, (typeof val === 'string' ? { type: val } : val));
        let res = { value: 0 };
        let currency = false;
        if (typeof item.type === 'function') {
            res = item.type(this.getValues(index, list), col, list, rawData);
            currency = true;
        }
        else {
            switch (item.type) {
                case 'count':
                    res.value = list.length;
                    break;
                case 'distinctCount':
                    res.value = this.getValues(index, list).filter((value, idx, self) => self.indexOf(value) === idx).length;
                    break;
                case 'sum':
                    res.value = this.toFixed(this.getSum(index, list), item.digits);
                    currency = true;
                    break;
                case 'average':
                    res.value = this.toFixed(this.getSum(index, list) / list.length, item.digits);
                    currency = true;
                    break;
                case 'max':
                    res.value = Math.max(...this.getValues(index, list));
                    currency = true;
                    break;
                case 'min':
                    res.value = Math.min(...this.getValues(index, list));
                    currency = true;
                    break;
            }
        }
        if (item.currency === true || (item.currency == null && currency === true)) {
            res.text = this.currencySrv.format(res.value, (_a = col.currency) === null || _a === void 0 ? void 0 : _a.format);
        }
        else {
            res.text = String(res.value);
        }
        return res;
    }
    toFixed(val, digits) {
        if (isNaN(val) || !isFinite(val)) {
            return 0;
        }
        return parseFloat(val.toFixed(digits));
    }
    getValues(index, list) {
        return list.map(i => i._values[index].org).map(i => (i === '' || i == null ? 0 : i));
    }
    getSum(index, list) {
        return this.getValues(index, list).reduce((p, i) => (p += parseFloat(String(i))), 0);
    }
}
STDataSource.ɵfac = function STDataSource_Factory(t) { return new (t || STDataSource)(ɵngcc0.ɵɵinject(ɵngcc3._HttpClient), ɵngcc0.ɵɵinject(ɵngcc3.DatePipe, 1), ɵngcc0.ɵɵinject(ɵngcc3.YNPipe, 1), ɵngcc0.ɵɵinject(ɵngcc4.DecimalPipe, 1), ɵngcc0.ɵɵinject(ɵngcc5.CurrencyService), ɵngcc0.ɵɵinject(ɵngcc1.DomSanitizer)); };
STDataSource.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: STDataSource, factory: STDataSource.ɵfac });
STDataSource.ctorParameters = () => [
    { type: _HttpClient },
    { type: DatePipe, decorators: [{ type: Host }] },
    { type: YNPipe, decorators: [{ type: Host }] },
    { type: DecimalPipe, decorators: [{ type: Host }] },
    { type: CurrencyService },
    { type: DomSanitizer }
];

class STExport {
    constructor(xlsxSrv) {
        this.xlsxSrv = xlsxSrv;
    }
    _stGet(item, col, index, colIndex) {
        const ret = { t: 's', v: '' };
        if (col.format) {
            ret.v = col.format(item, col, index);
        }
        else {
            const val = item._values ? item._values[colIndex].text : deepGet(item, col.index, '');
            ret.v = val;
            if (val != null) {
                switch (col.type) {
                    case 'currency':
                        ret.t = 'n';
                        break;
                    case 'date':
                        // Can't be a empty value, it will cause `#NULL!`
                        // See https://github.com/SheetJS/sheetjs/blob/master/docbits/52_datatype.md
                        if (`${val}`.length > 0) {
                            ret.t = 'd';
                            // Number Formats: https://github.com/SheetJS/sheetjs/blob/master/docbits/63_numfmt.md
                            ret.z = col.dateFormat;
                        }
                        break;
                    case 'yn':
                        const yn = col.yn;
                        ret.v = val === yn.truth ? yn.yes : yn.no;
                        break;
                }
            }
        }
        ret.v = ret.v || '';
        return ret;
    }
    genSheet(opt) {
        const sheets = {};
        const sheet = (sheets[opt.sheetname || 'Sheet1'] = {});
        const dataLen = opt.data.length;
        let validColCount = 0;
        let loseCount = 0;
        const columns = opt.columens;
        if (columns.findIndex(w => w._width != null) !== -1) {
            // wpx: width in screen pixels https://github.com/SheetJS/sheetjs#column-properties
            sheet['!cols'] = columns.map(col => ({ wpx: col._width }));
        }
        for (let colIdx = 0; colIdx < columns.length; colIdx++) {
            const col = columns[colIdx];
            if (col.exported === false || !col.index || !(!col.buttons || col.buttons.length === 0)) {
                ++loseCount;
                continue;
            }
            ++validColCount;
            const columnName = this.xlsxSrv.numberToSchema(colIdx + 1 - loseCount);
            sheet[`${columnName}1`] = {
                t: 's',
                v: typeof col.title === 'object' ? col.title.text : col.title
            };
            for (let dataIdx = 0; dataIdx < dataLen; dataIdx++) {
                sheet[`${columnName}${dataIdx + 2}`] = this._stGet(opt.data[dataIdx], col, dataIdx, colIdx);
            }
        }
        if (validColCount > 0 && dataLen > 0) {
            sheet['!ref'] = `A1:${this.xlsxSrv.numberToSchema(validColCount)}${dataLen + 1}`;
        }
        return sheets;
    }
    export(opt) {
        return __awaiter(this, void 0, void 0, function* () {
            const sheets = this.genSheet(opt);
            return this.xlsxSrv.export({
                sheets,
                filename: opt.filename,
                callback: opt.callback
            });
        });
    }
}
STExport.ɵfac = function STExport_Factory(t) { return new (t || STExport)(ɵngcc0.ɵɵinject(ɵngcc6.XlsxService, 8)); };
STExport.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: STExport, factory: STExport.ɵfac });
STExport.ctorParameters = () => [
    { type: XlsxService, decorators: [{ type: Optional }] }
];

class STWidgetHostDirective {
    constructor(stWidgetRegistry, viewContainerRef, componentFactoryResolver) {
        this.stWidgetRegistry = stWidgetRegistry;
        this.viewContainerRef = viewContainerRef;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    ngOnInit() {
        const widget = this.column.widget;
        const componentType = this.stWidgetRegistry.get(widget.type);
        const componentFactory = this.componentFactoryResolver.resolveComponentFactory(componentType);
        this.viewContainerRef.clear();
        const componentRef = this.viewContainerRef.createComponent(componentFactory);
        const { record, column } = this;
        const data = widget.params ? widget.params({ record, column }) : { record };
        Object.keys(data).forEach(key => {
            componentRef.instance[key] = data[key];
        });
    }
}
STWidgetHostDirective.ɵfac = function STWidgetHostDirective_Factory(t) { return new (t || STWidgetHostDirective)(ɵngcc0.ɵɵdirectiveInject(STWidgetRegistry), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver)); };
STWidgetHostDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: STWidgetHostDirective, selectors: [["", "st-widget-host", ""]], inputs: { record: "record", column: "column" } });
STWidgetHostDirective.ctorParameters = () => [
    { type: STWidgetRegistry },
    { type: ViewContainerRef },
    { type: ComponentFactoryResolver }
];
STWidgetHostDirective.propDecorators = {
    record: [{ type: Input }],
    column: [{ type: Input }]
};

const ST_DEFAULT_CONFIG = {
    pi: 1,
    ps: 10,
    size: 'default',
    responsive: true,
    responsiveHideHeaderFooter: false,
    req: {
        type: 'page',
        method: 'GET',
        allInBody: false,
        lazyLoad: false,
        reName: { pi: 'pi', ps: 'ps', skip: 'skip', limit: 'limit' }
    },
    res: {
        reName: { list: ['list'], total: ['total'] }
    },
    page: {
        front: true,
        zeroIndexed: false,
        position: 'bottom',
        placement: 'right',
        show: true,
        showSize: false,
        pageSizes: [10, 20, 30, 40, 50],
        showQuickJumper: false,
        total: true,
        toTop: true,
        toTopOffset: 100,
        itemRender: null,
        simple: false
    },
    modal: {
        paramsName: 'record',
        size: 'lg',
        exact: true
    },
    drawer: {
        paramsName: 'record',
        size: 'md',
        footer: true,
        footerHeight: 55
    },
    pop: {
        title: '确认删除吗？',
        trigger: 'click',
        placement: 'top'
    },
    btnIcon: {
        theme: 'outline',
        spin: false
    },
    noIndex: 1,
    expandRowByClick: false,
    expandAccordion: false,
    widthMode: {
        type: 'default',
        strictBehavior: 'truncate'
    },
    virtualItemSize: 54,
    virtualMaxBufferPx: 200,
    virtualMinBufferPx: 100,
    iifBehavior: 'hide',
    loadingDelay: 0,
    safeType: 'safeHtml',
    date: {
        format: `yyyy-MM-dd HH:mm`
    },
    yn: {
        truth: true,
        yes: '是',
        mode: 'icon'
    }
};

class STComponent {
    constructor(i18nSrv, cdr, el, exportSrv, doc, columnSource, dataSource, delonI18n, configSrv, cms) {
        this.cdr = cdr;
        this.el = el;
        this.exportSrv = exportSrv;
        this.doc = doc;
        this.columnSource = columnSource;
        this.dataSource = dataSource;
        this.delonI18n = delonI18n;
        this.cms = cms;
        this.destroy$ = new Subject();
        this.totalTpl = ``;
        this.customWidthConfig = false;
        this._widthConfig = [];
        this.locale = {};
        this._loading = false;
        this._data = [];
        this._statistical = {};
        this._isPagination = true;
        this._allChecked = false;
        this._allCheckedDisabled = false;
        this._indeterminate = false;
        this._headers = [];
        this._columns = [];
        this.contextmenuList = [];
        this.columns = [];
        this.ps = 10;
        this.pi = 1;
        this.total = 0;
        this.loading = null;
        this.loadingDelay = 0;
        this.bordered = false;
        this.showHeader = true;
        this.expandRowByClick = false;
        this.expandAccordion = false;
        this.responsive = true;
        this.error = new EventEmitter();
        this.change = new EventEmitter();
        this.virtualScroll = false;
        this.virtualItemSize = 54;
        this.virtualMaxBufferPx = 200;
        this.virtualMinBufferPx = 100;
        this.virtualForTrackBy = index => index;
        this.setCog(configSrv.merge('st', ST_DEFAULT_CONFIG));
        this.delonI18n.change.pipe(takeUntil(this.destroy$)).subscribe(() => {
            this.locale = this.delonI18n.getData('st');
            if (this._columns.length > 0) {
                this.updateTotalTpl();
                this.cd();
            }
        });
        i18nSrv.change
            .pipe(takeUntil(this.destroy$), filter(() => this._columns.length > 0))
            .subscribe(() => this.refreshColumns());
    }
    get req() {
        return this._req;
    }
    set req(value) {
        this._req = deepMergeKey({}, true, this.cog.req, value);
    }
    /** 返回体配置 */
    get res() {
        return this._res;
    }
    set res(value) {
        const item = (this._res = deepMergeKey({}, true, this.cog.res, value));
        const reName = item.reName;
        if (!Array.isArray(reName.list))
            reName.list = reName.list.split('.');
        if (!Array.isArray(reName.total))
            reName.total = reName.total.split('.');
        this._res = item;
    }
    get page() {
        return this._page;
    }
    set page(value) {
        this._page = Object.assign(Object.assign({}, this.cog.page), value);
        this.updateTotalTpl();
    }
    get multiSort() {
        return this._multiSort;
    }
    set multiSort(value) {
        if ((typeof value === 'boolean' && !toBoolean(value)) ||
            (typeof value === 'object' && Object.keys(value).length === 0)) {
            this._multiSort = undefined;
            return;
        }
        this._multiSort = Object.assign({}, (typeof value === 'object' ? value : {}));
    }
    set widthMode(value) {
        this._widthMode = Object.assign(Object.assign({}, this.cog.widthMode), value);
    }
    get widthMode() {
        return this._widthMode;
    }
    set widthConfig(val) {
        this._widthConfig = val;
        this.customWidthConfig = val && val.length > 0;
    }
    set resizable(val) {
        this._resizable = typeof val === 'object' ? val : { disabled: !toBoolean(val) };
    }
    /**
     * Get the number of the current page
     */
    get count() {
        return this._data.length;
    }
    /**
     * Get the data of the current page
     */
    get list() {
        return this._data;
    }
    setCog(cog) {
        const copyMultiSort = Object.assign({}, cog.multiSort);
        // Because multiSort.global will affect the result, it should be removed first, and multiSort will be operated again after processing.
        delete cog.multiSort;
        this.cog = cog;
        Object.assign(this, cog);
        if (copyMultiSort.global !== false) {
            this.multiSort = copyMultiSort;
        }
        this.columnSource.setCog(cog);
    }
    cd() {
        this.cdr.detectChanges();
        return this;
    }
    renderTotal(total, range) {
        return this.totalTpl
            ? this.totalTpl.replace('{{total}}', total).replace('{{range[0]}}', range[0]).replace('{{range[1]}}', range[1])
            : '';
    }
    changeEmit(type, data) {
        const res = {
            type,
            pi: this.pi,
            ps: this.ps,
            total: this.total
        };
        if (data != null) {
            res[type] = data;
        }
        this.change.emit(res);
    }
    // #region data
    /**
     * 获取过滤后所有数据
     * - 本地数据：包含排序、过滤后不分页数据
     * - 远程数据：不传递 `pi`、`ps` 两个参数
     */
    get filteredData() {
        return this.loadData({ paginator: false }).then(res => res.list);
    }
    updateTotalTpl() {
        const { total } = this.page;
        if (typeof total === 'string' && total.length) {
            this.totalTpl = total;
        }
        else if (toBoolean(total)) {
            this.totalTpl = this.locale.total;
        }
        else {
            this.totalTpl = '';
        }
    }
    setLoading(val) {
        if (this.loading == null) {
            this._loading = val;
            this.cdr.detectChanges();
        }
    }
    loadData(options) {
        const { pi, ps, data, req, res, page, total, singleSort, multiSort, rowClassName } = this;
        return new Promise((resolvePromise, rejectPromise) => {
            if (this.data$) {
                this.data$.unsubscribe();
            }
            this.data$ = this.dataSource
                .process(Object.assign({ pi,
                ps,
                total,
                data,
                req,
                res,
                page, columns: this._columns, singleSort,
                multiSort,
                rowClassName, paginator: true, customRequest: this.customRequest || this.cog.customRequest }, options))
                .pipe(takeUntil(this.destroy$))
                .subscribe(result => resolvePromise(result), error => {
                console.warn('st.loadDate', error);
                rejectPromise(error);
            });
        });
    }
    loadPageData() {
        return __awaiter(this, void 0, void 0, function* () {
            this.setLoading(true);
            try {
                const result = yield this.loadData();
                this.setLoading(false);
                if (typeof result.pi !== 'undefined') {
                    this.pi = result.pi;
                }
                if (typeof result.ps !== 'undefined') {
                    this.ps = result.ps;
                }
                if (typeof result.total !== 'undefined') {
                    this.total = result.total;
                }
                if (typeof result.pageShow !== 'undefined') {
                    this._isPagination = result.pageShow;
                }
                this._data = result.list;
                this._statistical = result.statistical;
                this.changeEmit('loaded', result.list);
                // Should be re-render in next tike when using virtual scroll
                // https://github.com/ng-alain/ng-alain/issues/1836
                if (this.cdkVirtualScrollViewport) {
                    Promise.resolve().then(() => this.cdkVirtualScrollViewport.checkViewportSize());
                }
                return this._refCheck();
            }
            catch (error) {
                this.setLoading(false);
                if (!this.destroy$.isStopped) {
                    this.cdr.detectChanges();
                    this.error.emit({ type: 'req', error });
                }
                return this;
            }
        });
    }
    /** 清空所有数据 */
    clear(cleanStatus = true) {
        if (cleanStatus) {
            this.clearStatus();
        }
        this._data = [];
        return this.cd();
    }
    /** 清空所有状态 */
    clearStatus() {
        return this.clearCheck().clearRadio().clearFilter().clearSort();
    }
    /**
     * 根据页码重新加载数据
     *
     * @param pi 指定当前页码，默认：`1`
     * @param extraParams 重新指定 `extraParams` 值
     * @param options 选项
     */
    load(pi = 1, extraParams, options) {
        if (pi !== -1)
            this.pi = pi;
        if (typeof extraParams !== 'undefined') {
            this.req.params = options && options.merge ? Object.assign(Object.assign({}, this.req.params), extraParams) : extraParams;
        }
        this._change('pi', options);
        return this;
    }
    /**
     * 重新刷新当前页
     *
     * @param extraParams 重新指定 `extraParams` 值
     */
    reload(extraParams, options) {
        return this.load(-1, extraParams, options);
    }
    /**
     * 重置且重新设置 `pi` 为 `1`，包含以下值：
     * - `check` 数据
     * - `radio` 数据
     * - `sort` 数据
     * - `fileter` 数据
     *
     * @param extraParams 重新指定 `extraParams` 值
     */
    reset(extraParams, options) {
        this.clearStatus().load(1, extraParams, options);
        return this;
    }
    _toTop(enforce) {
        var _a;
        if (!(enforce == null ? this.page.toTop : enforce))
            return;
        const el = this.el.nativeElement;
        el.scrollIntoView();
        // fix header height
        this.doc.documentElement.scrollTop -= this.page.toTopOffset;
        if (this.scroll) {
            if (this.cdkVirtualScrollViewport) {
                this.cdkVirtualScrollViewport.scrollTo({
                    top: 0,
                    left: 0
                });
            }
            else {
                (_a = el.querySelector('.ant-table-body, .ant-table-content')) === null || _a === void 0 ? void 0 : _a.scrollTo(0, 0);
            }
        }
    }
    _change(type, options) {
        if (type === 'pi' || (type === 'ps' && this.pi <= Math.ceil(this.total / this.ps))) {
            this.loadPageData().then(() => this._toTop(options === null || options === void 0 ? void 0 : options.toTop));
        }
        this.changeEmit(type);
    }
    closeOtherExpand(item) {
        if (this.expandAccordion === false)
            return;
        this._data.filter(i => i !== item).forEach(i => (i.expand = false));
    }
    _rowClick(e, item, index, dbl) {
        const el = e.target;
        if (el.nodeName === 'INPUT')
            return;
        const { expand, expandRowByClick } = this;
        if (!!expand && item.showExpand !== false && expandRowByClick) {
            item.expand = !item.expand;
            this.closeOtherExpand(item);
            this.changeEmit('expand', item);
            return;
        }
        const data = { e, item, index };
        if (dbl) {
            this.changeEmit('dblClick', data);
        }
        else {
            this._clickRowClassName(el, item, index);
            this.changeEmit('click', data);
        }
    }
    _clickRowClassName(el, item, index) {
        const cr = this.clickRowClassName;
        if (cr == null)
            return;
        const config = Object.assign({ exclusive: false }, (typeof cr === 'string' ? { fn: () => cr } : cr));
        const className = config.fn(item, index);
        const trEl = el.closest('tr');
        if (config.exclusive) {
            trEl.parentElement.querySelectorAll('tr').forEach((a) => a.classList.remove(className));
        }
        if (trEl.classList.contains(className)) {
            trEl.classList.remove(className);
        }
        else {
            trEl.classList.add(className);
        }
    }
    _expandChange(item, expand) {
        item.expand = expand;
        this.closeOtherExpand(item);
        this.changeEmit('expand', item);
    }
    _stopPropagation(ev) {
        ev.stopPropagation();
    }
    /**
     * Remove a row in the table, like this:
     *
     * ```
     * this.st.removeRow(0)
     * this.st.removeRow(stDataItem)
     * ```
     */
    removeRow(data) {
        if (typeof data === 'number') {
            this._data.splice(data, 1);
        }
        else {
            if (!Array.isArray(data)) {
                data = [data];
            }
            data
                .map(item => this._data.indexOf(item))
                .filter(pos => pos !== -1)
                .forEach(pos => this._data.splice(pos, 1));
        }
        // recalculate no
        this._columns
            .filter(w => w.type === 'no')
            .forEach(c => this._data.forEach((i, idx) => {
            const text = `${this.dataSource.getNoIndex(i, c, idx)}`;
            i._values[c.__point] = { text, _text: text, org: idx, safeType: 'text' };
        }));
        return this.cd();
    }
    /**
     * Sets the row value for the `index` in the table, like this:
     *
     * - `optinos.refreshSchema` Whether to refresh of st schemas
     * - `optinos.emitReload` Whether to trigger a reload http request when data is url
     *
     * ```
     * this.st.setRow(0, { price: 100 })
     * this.st.setRow(0, { price: 100, name: 'asdf' })
     * this.st.setRow(item, { price: 100 })
     * ```
     */
    setRow(index, item, options) {
        options = Object.assign({ refreshSchema: false, emitReload: false }, options);
        if (typeof index !== 'number') {
            index = this._data.indexOf(index);
        }
        this._data[index] = deepMergeKey(this._data[index], false, item);
        this.optimizeData();
        if (options.refreshSchema) {
            this.resetColumns({ emitReload: options.emitReload });
            return this;
        }
        this.cdr.detectChanges();
        return this;
    }
    // #endregion
    // #region sort
    sort(col, idx, value) {
        if (this.multiSort) {
            col._sort.default = value;
            col._sort.tick = this.dataSource.nextSortTick;
        }
        else {
            this._columns.forEach((item, index) => (item._sort.default = index === idx ? value : null));
        }
        this.cdr.detectChanges();
        this.loadPageData();
        const res = {
            value,
            map: this.dataSource.getReqSortMap(this.singleSort, this.multiSort, this._columns),
            column: col
        };
        this.changeEmit('sort', res);
    }
    clearSort() {
        this._columns.forEach(item => (item._sort.default = null));
        return this;
    }
    // #endregion
    // #region filter
    _handleFilter(col, confirm) {
        if (!confirm) {
            this.columnSource.cleanFilter(col);
        }
        // 过滤表示一种数据的变化应重置页码为 `1`
        this.pi = 1;
        this.columnSource.updateDefault(col.filter);
        this.loadPageData();
        this.changeEmit('filter', col);
    }
    handleFilterNotify(value) {
        this.changeEmit('filterChange', value);
    }
    clearFilter() {
        this._columns.filter(w => w.filter && w.filter.default === true).forEach(col => this.columnSource.cleanFilter(col));
        return this;
    }
    // #endregion
    // #region checkbox
    /** 清除所有 `checkbox` */
    clearCheck() {
        return this._checkAll(false);
    }
    _refCheck() {
        const validData = this._data.filter(w => !w.disabled);
        const checkedList = validData.filter(w => w.checked === true);
        this._allChecked = checkedList.length > 0 && checkedList.length === validData.length;
        const allUnChecked = validData.every(value => !value.checked);
        this._indeterminate = !this._allChecked && !allUnChecked;
        this._allCheckedDisabled = this._data.length === this._data.filter(w => w.disabled).length;
        return this.cd();
    }
    _checkAll(checked) {
        checked = typeof checked === 'undefined' ? this._allChecked : checked;
        this._data.filter(w => !w.disabled).forEach(i => (i.checked = checked));
        return this._refCheck()._checkNotify();
    }
    _rowSelection(row) {
        row.select(this._data);
        return this._refCheck()._checkNotify();
    }
    _checkNotify() {
        const res = this._data.filter(w => !w.disabled && w.checked === true);
        this.changeEmit('checkbox', res);
        return this;
    }
    // #endregion
    // #region radio
    /** 清除所有 `radio` */
    clearRadio() {
        this._data.filter(w => w.checked).forEach(item => (item.checked = false));
        this.changeEmit('radio', null);
        return this;
    }
    // #endregion
    _handleTd(ev) {
        switch (ev.type) {
            case 'checkbox':
                this._refCheck()._checkNotify();
                break;
            case 'radio':
                this.changeEmit('radio', ev.item);
                break;
        }
    }
    // #region export
    /**
     * 导出当前页，确保已经注册 `XlsxModule`
     *
     * @param newData 重新指定数据；若为 `true` 表示使用 `filteredData` 数据
     * @param opt 额外参数
     */
    export(newData, opt) {
        const data = Array.isArray(newData)
            ? this.dataSource.optimizeData({ columns: this._columns, result: newData })
            : this._data;
        (newData === true ? from(this.filteredData) : of(data)).subscribe((res) => this.exportSrv.export(Object.assign(Object.assign({ columens: this._columns }, opt), { data: res })));
    }
    // #endregion
    // #region resizable
    colResize({ width }, column) {
        column.width = `${width}px`;
        this.changeEmit('resize', column);
    }
    // #endregion
    // #region contextmenu
    onContextmenu(event) {
        if (!this.contextmenu) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        const colEl = event.target.closest('[data-col-index]');
        if (!colEl) {
            return;
        }
        const colIndex = Number(colEl.dataset.colIndex);
        const rowIndex = Number(colEl.closest('tr').dataset.index);
        const isTitle = isNaN(rowIndex);
        const obs$ = this.contextmenu({
            event,
            type: isTitle ? 'head' : 'body',
            rowIndex: isTitle ? null : rowIndex,
            colIndex,
            data: isTitle ? null : this.list[rowIndex],
            column: this._columns[colIndex]
        });
        (isObservable(obs$) ? obs$ : of(obs$))
            .pipe(takeUntil(this.destroy$), filter(res => res.length > 0))
            .subscribe(res => {
            this.contextmenuList = res.map(i => {
                if (!Array.isArray(i.children)) {
                    i.children = [];
                }
                return i;
            });
            this.cdr.detectChanges();
            this.cms.create(event, this.contextmenuTpl);
        });
    }
    // #endregion
    get cdkVirtualScrollViewport() {
        return this.orgTable.cdkVirtualScrollViewport;
    }
    resetColumns(options) {
        options = Object.assign({ emitReload: true, preClearData: false }, options);
        if (typeof options.columns !== 'undefined') {
            this.columns = options.columns;
        }
        if (typeof options.pi !== 'undefined') {
            this.pi = options.pi;
        }
        if (typeof options.ps !== 'undefined') {
            this.ps = options.ps;
        }
        if (options.emitReload) {
            // Should clean data, Because of changing columns may cause inaccurate data
            options.preClearData = true;
        }
        if (options.preClearData) {
            this._data = [];
        }
        this.refreshColumns();
        if (options.emitReload) {
            return this.loadPageData();
        }
        else {
            this.cd();
            return Promise.resolve(this);
        }
    }
    refreshColumns() {
        const res = this.columnSource.process(this.columns, {
            widthMode: this.widthMode,
            resizable: this._resizable,
            safeType: this.cog.safeType
        });
        this._columns = res.columns;
        this._headers = res.headers;
        if (this.customWidthConfig === false && res.headerWidths != null) {
            this._widthConfig = res.headerWidths;
        }
        return this;
    }
    optimizeData() {
        this._data = this.dataSource.optimizeData({
            columns: this._columns,
            result: this._data,
            rowClassName: this.rowClassName
        });
    }
    /**
     * Return pure data, `st` internally maintains a set of data for caching, this part of data may affect the backend
     *
     * 返回纯净数据，`st` 内部会维护一组用于缓存的数据，这部分数据可能会影响后端
     */
    pureItem(itemOrIndex) {
        if (typeof itemOrIndex === 'number') {
            itemOrIndex = this._data[itemOrIndex];
        }
        if (!itemOrIndex) {
            return null;
        }
        const copyItem = deepCopy(itemOrIndex);
        delete copyItem._values;
        return copyItem;
    }
    ngAfterViewInit() {
        this.columnSource.restoreAllRender(this._columns);
    }
    ngOnChanges(changes) {
        if (changes.columns) {
            this.refreshColumns().optimizeData();
        }
        const changeData = changes.data;
        if (changeData && changeData.currentValue && !(this.req.lazyLoad && changeData.firstChange)) {
            this.loadPageData();
        }
        if (changes.loading) {
            this._loading = changes.loading.currentValue;
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
STComponent.ɵfac = function STComponent_Factory(t) { return new (t || STComponent)(ɵngcc0.ɵɵdirectiveInject(ALAIN_I18N_TOKEN, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(STExport), ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(STColumnSource), ɵngcc0.ɵɵdirectiveInject(STDataSource), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DelonLocaleService), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.AlainConfigService), ɵngcc0.ɵɵdirectiveInject(ɵngcc8.NzContextMenuService)); };
STComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: STComponent, selectors: [["st"]], viewQuery: function STComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
        ɵngcc0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.orgTable = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contextmenuTpl = _t.first);
    } }, hostVars: 12, hostBindings: function STComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("st", true)("st__p-left", ctx.page.placement === "left")("st__p-center", ctx.page.placement === "center")("st__width-strict", ctx.widthMode.type === "strict")("ant-table-rep", ctx.responsive)("ant-table-rep__hide-header-footer", ctx.responsiveHideHeaderFooter);
    } }, inputs: { columns: "columns", ps: "ps", pi: "pi", total: "total", loading: "loading", loadingDelay: "loadingDelay", bordered: "bordered", showHeader: "showHeader", expandRowByClick: "expandRowByClick", expandAccordion: "expandAccordion", responsive: "responsive", virtualScroll: "virtualScroll", virtualItemSize: "virtualItemSize", virtualMaxBufferPx: "virtualMaxBufferPx", virtualMinBufferPx: "virtualMinBufferPx", virtualForTrackBy: "virtualForTrackBy", req: "req", res: "res", page: "page", multiSort: "multiSort", widthMode: "widthMode", widthConfig: "widthConfig", resizable: "resizable", data: "data", contextmenu: "contextmenu", loadingIndicator: "loadingIndicator", size: "size", scroll: "scroll", singleSort: "singleSort", rowClassName: "rowClassName", clickRowClassName: "clickRowClassName", header: "header", footer: "footer", bodyHeader: "bodyHeader", body: "body", expand: "expand", noResult: "noResult", responsiveHideHeaderFooter: "responsiveHideHeaderFooter", customRequest: "customRequest" }, outputs: { error: "error", change: "change" }, exportAs: ["st"], features: [ɵngcc0.ɵɵProvidersFeature([STDataSource, STRowSource, STColumnSource, STExport, DatePipe, YNPipe, DecimalPipe]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 20, vars: 34, consts: [["titleTpl", ""], ["chkAllTpl", ""], [3, "nzData", "nzPageIndex", "nzPageSize", "nzTotal", "nzShowPagination", "nzFrontPagination", "nzBordered", "nzSize", "nzLoading", "nzLoadingDelay", "nzLoadingIndicator", "nzTitle", "nzFooter", "nzScroll", "nzVirtualItemSize", "nzVirtualMaxBufferPx", "nzVirtualMinBufferPx", "nzVirtualForTrackBy", "nzNoResult", "nzPageSizeOptions", "nzShowQuickJumper", "nzShowSizeChanger", "nzPaginationPosition", "nzPaginationType", "nzItemRender", "nzSimple", "nzShowTotal", "nzWidthConfig", "nzPageIndexChange", "nzPageSizeChange", "contextmenu"], ["table", ""], [4, "ngIf"], [1, "st__body"], ["bodyTpl", ""], ["totalTpl", ""], ["contextmenuTpl", "nzDropdownMenu"], ["nz-menu", "", 1, "st__contextmenu"], [4, "ngFor", "ngForOf"], [3, "innerHTML"], ["class", "st__head-optional", 3, "innerHTML", 4, "ngIf"], ["class", "st__head-tip", "nz-tooltip", "", "nz-icon", "", "nzType", "question-circle", 3, "nzTooltipTitle", 4, "ngIf"], [1, "st__head-optional", 3, "innerHTML"], ["nz-tooltip", "", "nz-icon", "", "nzType", "question-circle", 1, "st__head-tip", 3, "nzTooltipTitle"], ["nz-checkbox", "", 1, "st__checkall", 3, "nzDisabled", "ngModel", "nzIndeterminate", "ngModelChange"], ["nzWidth", "50px", 3, "rowSpan", 4, "ngIf"], ["nzWidth", "50px", 3, "rowSpan"], ["nz-resizable", "", 3, "colSpan", "rowSpan", "nzWidth", "nzLeft", "nzRight", "ngClass", "nzShowSort", "nzSortOrder", "nzCustomFilter", "nzDisabled", "nzMaxWidth", "nzMinWidth", "nzBounds", "nzPreview", "nzSortOrderChange", "nzResizeEnd", 4, "let"], ["nz-resizable", "", 3, "colSpan", "rowSpan", "nzWidth", "nzLeft", "nzRight", "ngClass", "nzShowSort", "nzSortOrder", "nzCustomFilter", "nzDisabled", "nzMaxWidth", "nzMinWidth", "nzBounds", "nzPreview", "nzSortOrderChange", "nzResizeEnd"], ["nzDirection", "right", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["renderTitle", ""], [4, "ngIf", "ngIfElse"], ["nzDirection", "right"], [3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngSwitchDefault"], ["class", "ant-table-selection", 4, "ngIf"], [1, "ant-table-selection"], ["class", "ant-table-selection-extra", 4, "ngIf"], ["selectionMenu", "nzDropdownMenu"], ["nz-menu", "", 1, "ant-table-selection-menu"], ["nz-menu-item", "", 3, "innerHTML", "click", 4, "ngFor", "ngForOf"], [1, "ant-table-selection-extra"], ["nz-dropdown", "", "nzPlacement", "bottomLeft", 1, "ant-table-selection-down", "st__checkall-selection", 3, "nzDropdownMenu"], ["nz-icon", "", "nzType", "down"], ["nz-menu-item", "", 3, "innerHTML", "click"], ["nz-th-extra", "", 3, "col", "f", "locale", "n", "handle"], [3, "ngClass", "click", "dblclick"], ["nzWidth", "50px", 3, "nzShowExpand", "nzExpand", "nzExpandChange", "click", 4, "ngIf"], [3, "nzLeft", "nzRight", "ngClass", 4, "ngFor", "ngForOf"], [3, "nzExpand"], ["nzWidth", "50px", 3, "nzShowExpand", "nzExpand", "nzExpandChange", "click"], [3, "nzLeft", "nzRight", "ngClass"], ["class", "ant-table-rep__title", 4, "ngIf"], [3, "data", "i", "index", "c", "cIdx", "n"], [1, "ant-table-rep__title"], ["nz-virtual-scroll", ""], ["nz-menu-item", "", 3, "innerHTML", "click", 4, "ngIf"], ["nz-submenu", "", 3, "nzTitle", 4, "ngIf"], ["nz-submenu", "", 3, "nzTitle"]], template: function STComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, STComponent_ng_template_0_Template, 3, 3, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(2, STComponent_ng_template_2_Template, 1, 5, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementStart(4, "nz-table", 2, 3);
        ɵngcc0.ɵɵlistener("nzPageIndexChange", function STComponent_Template_nz_table_nzPageIndexChange_4_listener($event) { return ctx.pi = $event; })("nzPageIndexChange", function STComponent_Template_nz_table_nzPageIndexChange_4_listener() { return ctx._change("pi"); })("nzPageSizeChange", function STComponent_Template_nz_table_nzPageSizeChange_4_listener($event) { return ctx.ps = $event; })("nzPageSizeChange", function STComponent_Template_nz_table_nzPageSizeChange_4_listener() { return ctx._change("ps"); })("contextmenu", function STComponent_Template_nz_table_contextmenu_4_listener($event) { return ctx.onContextmenu($event); });
        ɵngcc0.ɵɵtemplate(6, STComponent_thead_6_Template, 2, 1, "thead", 4);
        ɵngcc0.ɵɵelementStart(7, "tbody", 5);
        ɵngcc0.ɵɵtemplate(8, STComponent_ng_container_8_Template, 2, 4, "ng-container", 4);
        ɵngcc0.ɵɵtemplate(9, STComponent_ng_template_9_Template, 5, 10, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(11, STComponent_ng_container_11_Template, 2, 1, "ng-container", 4);
        ɵngcc0.ɵɵtemplate(12, STComponent_ng_container_12_Template, 2, 0, "ng-container", 4);
        ɵngcc0.ɵɵtemplate(13, STComponent_ng_container_13_Template, 2, 4, "ng-container", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(14, STComponent_ng_template_14_Template, 1, 1, "ng-template", null, 7, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "nz-dropdown-menu", null, 8);
        ɵngcc0.ɵɵelementStart(18, "ul", 9);
        ɵngcc0.ɵɵtemplate(19, STComponent_ng_container_19_Template, 3, 2, "ng-container", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r12 = ɵngcc0.ɵɵreference(15);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("nzData", ctx._data)("nzPageIndex", ctx.pi)("nzPageSize", ctx.ps)("nzTotal", ctx.total)("nzShowPagination", ctx._isPagination)("nzFrontPagination", false)("nzBordered", ctx.bordered)("nzSize", ctx.size)("nzLoading", ctx._loading)("nzLoadingDelay", ctx.loadingDelay)("nzLoadingIndicator", ctx.loadingIndicator)("nzTitle", ctx.header)("nzFooter", ctx.footer)("nzScroll", ctx.scroll)("nzVirtualItemSize", ctx.virtualItemSize)("nzVirtualMaxBufferPx", ctx.virtualMaxBufferPx)("nzVirtualMinBufferPx", ctx.virtualMinBufferPx)("nzVirtualForTrackBy", ctx.virtualForTrackBy)("nzNoResult", ctx.noResult)("nzPageSizeOptions", ctx.page.pageSizes)("nzShowQuickJumper", ctx.page.showQuickJumper)("nzShowSizeChanger", ctx.page.showSize)("nzPaginationPosition", ctx.page.position)("nzPaginationType", ctx.page.type)("nzItemRender", ctx.page.itemRender)("nzSimple", ctx.page.simple)("nzShowTotal", _r12)("nzWidthConfig", ctx._widthConfig);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showHeader);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx._loading);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.virtualScroll);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.virtualScroll);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx._loading);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.contextmenuList);
    } }, directives: function () { return [ɵngcc9.NzTableComponent, ɵngcc4.NgIf, ɵngcc9.NzTbodyComponent, ɵngcc8.NzDropdownMenuComponent, ɵngcc10.NzMenuDirective, ɵngcc4.NgForOf, ɵngcc11.NzTooltipDirective, ɵngcc12.NzIconDirective, ɵngcc13.NzCheckboxComponent, ɵngcc14.NgControlStatus, ɵngcc14.NgModel, ɵngcc9.NzTheadComponent, ɵngcc9.NzTrDirective, ɵngcc9.NzTableCellDirective, ɵngcc9.NzThMeasureDirective, ɵngcc15.LetDirective, ɵngcc9.NzCellFixedDirective, ɵngcc9.NzThAddOnComponent, ɵngcc16.NzResizableDirective, ɵngcc4.NgClass, ɵngcc4.NgTemplateOutlet, ɵngcc16.NzResizeHandleComponent, ɵngcc4.NgSwitch, ɵngcc4.NgSwitchCase, ɵngcc4.NgSwitchDefault, ɵngcc8.NzDropDownDirective, ɵngcc10.NzMenuItemDirective, STFilterComponent, ɵngcc9.NzTrExpandDirective, ɵngcc9.NzTableFixedRowComponent, ɵngcc9.NzTdAddOnComponent, STTdComponent, ɵngcc9.NzTableVirtualScrollDirective, ɵngcc10.NzSubMenuComponent]; }, encapsulation: 2, changeDetection: 0 });
STComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [ALAIN_I18N_TOKEN,] }] },
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: STExport },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: STColumnSource },
    { type: STDataSource },
    { type: DelonLocaleService },
    { type: AlainConfigService },
    { type: NzContextMenuService }
];
STComponent.propDecorators = {
    orgTable: [{ type: ViewChild, args: ['table',] }],
    contextmenuTpl: [{ type: ViewChild, args: ['contextmenuTpl',] }],
    req: [{ type: Input }],
    res: [{ type: Input }],
    page: [{ type: Input }],
    data: [{ type: Input }],
    columns: [{ type: Input }],
    contextmenu: [{ type: Input }],
    ps: [{ type: Input }],
    pi: [{ type: Input }],
    total: [{ type: Input }],
    loading: [{ type: Input }],
    loadingDelay: [{ type: Input }],
    loadingIndicator: [{ type: Input }],
    bordered: [{ type: Input }],
    size: [{ type: Input }],
    scroll: [{ type: Input }],
    singleSort: [{ type: Input }],
    multiSort: [{ type: Input }],
    rowClassName: [{ type: Input }],
    clickRowClassName: [{ type: Input }],
    widthMode: [{ type: Input }],
    widthConfig: [{ type: Input }],
    resizable: [{ type: Input }],
    header: [{ type: Input }],
    showHeader: [{ type: Input }],
    footer: [{ type: Input }],
    bodyHeader: [{ type: Input }],
    body: [{ type: Input }],
    expandRowByClick: [{ type: Input }],
    expandAccordion: [{ type: Input }],
    expand: [{ type: Input }],
    noResult: [{ type: Input }],
    responsive: [{ type: Input }],
    responsiveHideHeaderFooter: [{ type: Input }],
    error: [{ type: Output }],
    change: [{ type: Output }],
    virtualScroll: [{ type: Input }],
    virtualItemSize: [{ type: Input }],
    virtualMaxBufferPx: [{ type: Input }],
    virtualMinBufferPx: [{ type: Input }],
    customRequest: [{ type: Input }],
    virtualForTrackBy: [{ type: Input }]
};
__decorate([
    InputNumber()
], STComponent.prototype, "ps", void 0);
__decorate([
    InputNumber()
], STComponent.prototype, "pi", void 0);
__decorate([
    InputNumber()
], STComponent.prototype, "total", void 0);
__decorate([
    InputNumber()
], STComponent.prototype, "loadingDelay", void 0);
__decorate([
    InputBoolean()
], STComponent.prototype, "bordered", void 0);
__decorate([
    InputBoolean()
], STComponent.prototype, "showHeader", void 0);
__decorate([
    InputBoolean()
], STComponent.prototype, "expandRowByClick", void 0);
__decorate([
    InputBoolean()
], STComponent.prototype, "expandAccordion", void 0);
__decorate([
    InputBoolean()
], STComponent.prototype, "responsive", void 0);
__decorate([
    InputBoolean()
], STComponent.prototype, "responsiveHideHeaderFooter", void 0);
__decorate([
    InputBoolean()
], STComponent.prototype, "virtualScroll", void 0);
__decorate([
    InputNumber()
], STComponent.prototype, "virtualItemSize", void 0);
__decorate([
    InputNumber()
], STComponent.prototype, "virtualMaxBufferPx", void 0);
__decorate([
    InputNumber()
], STComponent.prototype, "virtualMinBufferPx", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STRowSource, [{
        type: Injectable
    }], function () { return []; }, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STRowDirective, [{
        type: Directive,
        args: [{ selector: '[st-row]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }, { type: STRowSource, decorators: [{
                type: Host
            }] }]; }, { id: [{
            type: Input,
            args: ['st-row']
        }], type: [{
            type: Input
        }] }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STWidgetRegistry, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STColumnSource, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DomSanitizer }, { type: STRowSource, decorators: [{
                type: Host
            }] }, { type: ɵngcc2.ACLService, decorators: [{
                type: Optional
            }] }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [ALAIN_I18N_TOKEN]
            }] }, { type: STWidgetRegistry }]; }, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STDataSource, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc3._HttpClient }, { type: ɵngcc3.DatePipe, decorators: [{
                type: Host
            }] }, { type: ɵngcc3.YNPipe, decorators: [{
                type: Host
            }] }, { type: ɵngcc4.DecimalPipe, decorators: [{
                type: Host
            }] }, { type: ɵngcc5.CurrencyService }, { type: ɵngcc1.DomSanitizer }]; }, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STExport, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc6.XlsxService, decorators: [{
                type: Optional
            }] }]; }, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STWidgetHostDirective, [{
        type: Directive,
        args: [{ selector: '[st-widget-host]' }]
    }], function () { return [{ type: STWidgetRegistry }, { type: ɵngcc0.ViewContainerRef }, { type: ɵngcc0.ComponentFactoryResolver }]; }, { record: [{
            type: Input
        }], column: [{
            type: Input
        }] }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STComponent, [{
        type: Component,
        args: [{
                selector: 'st',
                exportAs: 'st',
                template: "<ng-template #titleTpl let-i>\n  <span [innerHTML]=\"i._text\"></span>\n  <small *ngIf=\"i.optional\" class=\"st__head-optional\" [innerHTML]=\"i.optional\"></small>\n  <i\n    *ngIf=\"i.optionalHelp\"\n    class=\"st__head-tip\"\n    nz-tooltip\n    [nzTooltipTitle]=\"i.optionalHelp\"\n    nz-icon\n    nzType=\"question-circle\"\n  ></i>\n</ng-template>\n<ng-template #chkAllTpl let-custom>\n  <label\n    nz-checkbox\n    class=\"st__checkall\"\n    [nzDisabled]=\"_allCheckedDisabled\"\n    [(ngModel)]=\"_allChecked\"\n    [nzIndeterminate]=\"_indeterminate\"\n    (ngModelChange)=\"_checkAll()\"\n    [class.ant-table-selection-select-all-custom]=\"custom\"\n  ></label>\n</ng-template>\n<nz-table\n  #table\n  [nzData]=\"_data\"\n  [(nzPageIndex)]=\"pi\"\n  (nzPageIndexChange)=\"_change('pi')\"\n  [(nzPageSize)]=\"ps\"\n  (nzPageSizeChange)=\"_change('ps')\"\n  [nzTotal]=\"total\"\n  [nzShowPagination]=\"_isPagination\"\n  [nzFrontPagination]=\"false\"\n  [nzBordered]=\"bordered\"\n  [nzSize]=\"size\"\n  [nzLoading]=\"_loading\"\n  [nzLoadingDelay]=\"loadingDelay\"\n  [nzLoadingIndicator]=\"loadingIndicator\"\n  [nzTitle]=\"header!\"\n  [nzFooter]=\"footer!\"\n  [nzScroll]=\"scroll\"\n  [nzVirtualItemSize]=\"virtualItemSize\"\n  [nzVirtualMaxBufferPx]=\"virtualMaxBufferPx\"\n  [nzVirtualMinBufferPx]=\"virtualMinBufferPx\"\n  [nzVirtualForTrackBy]=\"virtualForTrackBy\"\n  [nzNoResult]=\"noResult!\"\n  [nzPageSizeOptions]=\"page.pageSizes!\"\n  [nzShowQuickJumper]=\"page.showQuickJumper\"\n  [nzShowSizeChanger]=\"page.showSize\"\n  [nzPaginationPosition]=\"page.position!\"\n  [nzPaginationType]=\"page.type!\"\n  [nzItemRender]=\"page.itemRender!\"\n  [nzSimple]=\"page.simple\"\n  [nzShowTotal]=\"totalTpl\"\n  [nzWidthConfig]=\"_widthConfig\"\n  (contextmenu)=\"onContextmenu($event)\"\n>\n  <thead *ngIf=\"showHeader\">\n    <tr *ngFor=\"let row of _headers; let rowFirst = first\">\n      <th *ngIf=\"rowFirst && expand\" nzWidth=\"50px\" [rowSpan]=\"_headers.length\"></th>\n      <ng-container *ngFor=\"let h of row; let index = index; let last = last\">\n        <th\n          *let=\"h.column as _c\"\n          [colSpan]=\"h.colSpan\"\n          [rowSpan]=\"h.rowSpan\"\n          [nzWidth]=\"$any(_c).width\"\n          [nzLeft]=\"_c._left!\"\n          [nzRight]=\"_c._right!\"\n          [ngClass]=\"_c.className!\"\n          [attr.data-col]=\"_c.indexKey\"\n          [attr.data-col-index]=\"index\"\n          [nzShowSort]=\"_c._sort.enabled\"\n          [nzSortOrder]=\"$any(_c)._sort.default\"\n          (nzSortOrderChange)=\"sort(_c, index, $event)\"\n          [nzCustomFilter]=\"$any(_c).filter\"\n          nz-resizable\n          [nzDisabled]=\"last || $any(_c).resizable.disabled\"\n          [nzMaxWidth]=\"$any(_c).resizable.maxWidth\"\n          [nzMinWidth]=\"$any(_c).resizable.minWidth\"\n          [nzBounds]=\"$any(_c).resizable.bounds\"\n          [nzPreview]=\"$any(_c).resizable.preview\"\n          (nzResizeEnd)=\"colResize($event, _c)\"\n        >\n          <nz-resize-handle *ngIf=\"$any(!last && !$any(_c).resizable.disabled)\" nzDirection=\"right\">\n            <i></i>\n          </nz-resize-handle>\n          <ng-template\n            #renderTitle\n            [ngTemplateOutlet]=\"_c.__renderTitle!\"\n            [ngTemplateOutletContext]=\"{ $implicit: h.column, index: index }\"\n          ></ng-template>\n          <ng-container *ngIf=\"!_c.__renderTitle; else renderTitle\">\n            <ng-container [ngSwitch]=\"_c.type\">\n              <ng-container *ngSwitchCase=\"'checkbox'\">\n                <ng-container *ngIf=\"_c.selections!.length === 0\">\n                  <ng-template [ngTemplateOutlet]=\"chkAllTpl\" [ngTemplateOutletContext]=\"{ $implicit: false }\">\n                  </ng-template>\n                </ng-container>\n                <div *ngIf=\"_c.selections!.length > 0\" class=\"ant-table-selection\">\n                  <ng-template [ngTemplateOutlet]=\"chkAllTpl\" [ngTemplateOutletContext]=\"{ $implicit: true }\">\n                  </ng-template>\n                  <div *ngIf=\"_c.selections!.length\" class=\"ant-table-selection-extra\">\n                    <div\n                      nz-dropdown\n                      nzPlacement=\"bottomLeft\"\n                      [nzDropdownMenu]=\"selectionMenu\"\n                      class=\"ant-table-selection-down st__checkall-selection\"\n                    >\n                      <i nz-icon nzType=\"down\"></i>\n                    </div>\n                  </div>\n                  <nz-dropdown-menu #selectionMenu=\"nzDropdownMenu\">\n                    <ul nz-menu class=\"ant-table-selection-menu\">\n                      <li\n                        nz-menu-item\n                        *ngFor=\"let rw of _c.selections\"\n                        (click)=\"_rowSelection(rw)\"\n                        [innerHTML]=\"rw.text\"\n                      ></li>\n                    </ul>\n                  </nz-dropdown-menu>\n                </div>\n              </ng-container>\n              <ng-container *ngSwitchDefault>\n                <ng-template\n                  [ngTemplateOutlet]=\"titleTpl\"\n                  [ngTemplateOutletContext]=\"{ $implicit: _c.title }\"\n                ></ng-template>\n              </ng-container>\n            </ng-container>\n          </ng-container>\n          <ng-container *ngIf=\"_c.filter\">\n            <st-filter\n              nz-th-extra\n              [col]=\"h.column\"\n              [f]=\"_c.filter\"\n              [locale]=\"locale\"\n              (n)=\"handleFilterNotify($event)\"\n              (handle)=\"_handleFilter(_c, $event)\"\n            ></st-filter>\n          </ng-container>\n        </th>\n      </ng-container>\n    </tr>\n  </thead>\n  <tbody class=\"st__body\">\n    <ng-container *ngIf=\"!_loading\">\n      <ng-template\n        [ngTemplateOutlet]=\"bodyHeader!\"\n        [ngTemplateOutletContext]=\"{ $implicit: _statistical }\"\n      ></ng-template>\n    </ng-container>\n    <ng-template #bodyTpl let-i let-index=\"index\">\n      <tr\n        [attr.data-index]=\"index\"\n        (click)=\"_rowClick($event, i, index, false)\"\n        (dblclick)=\"_rowClick($event, i, index, true)\"\n        [ngClass]=\"i._rowClassName\"\n      >\n        <td\n          *ngIf=\"expand\"\n          [nzShowExpand]=\"expand && i.showExpand !== false\"\n          [nzExpand]=\"i.expand\"\n          (nzExpandChange)=\"_expandChange(i, $event)\"\n          (click)=\"_stopPropagation($event)\"\n          nzWidth=\"50px\"\n        ></td>\n        <td\n          *ngFor=\"let c of _columns; let cIdx = index\"\n          [nzLeft]=\"!!c._left\"\n          [nzRight]=\"!!c._right\"\n          [attr.data-col-index]=\"cIdx\"\n          [ngClass]=\"c._className!\"\n          [attr.colspan]=\"c.colSpan\"\n        >\n          <span *ngIf=\"responsive\" class=\"ant-table-rep__title\">\n            <ng-template [ngTemplateOutlet]=\"titleTpl\" [ngTemplateOutletContext]=\"{ $implicit: c.title }\"></ng-template>\n          </span>\n          <st-td [data]=\"_data\" [i]=\"i\" [index]=\"index\" [c]=\"c\" [cIdx]=\"cIdx\" (n)=\"_handleTd($event)\"></st-td>\n        </td>\n      </tr>\n      <tr [nzExpand]=\"i.expand\">\n        <ng-template\n          [ngTemplateOutlet]=\"expand\"\n          [ngTemplateOutletContext]=\"{ $implicit: i, index: index }\"\n        ></ng-template>\n      </tr>\n    </ng-template>\n    <ng-container *ngIf=\"!virtualScroll\">\n      <ng-container *ngFor=\"let i of _data; let index = index\">\n        <ng-template [ngTemplateOutlet]=\"bodyTpl\" [ngTemplateOutletContext]=\"{ $implicit: i, index: index }\">\n        </ng-template>\n      </ng-container>\n    </ng-container>\n    <ng-container *ngIf=\"virtualScroll\">\n      <ng-template nz-virtual-scroll let-i let-index=\"index\">\n        <ng-template [ngTemplateOutlet]=\"bodyTpl\" [ngTemplateOutletContext]=\"{ $implicit: i, index: index }\">\n        </ng-template>\n      </ng-template>\n    </ng-container>\n    <ng-container *ngIf=\"!_loading\">\n      <ng-template [ngTemplateOutlet]=\"body!\" [ngTemplateOutletContext]=\"{ $implicit: _statistical }\"></ng-template>\n    </ng-container>\n  </tbody>\n  <ng-template #totalTpl let-range=\"range\" let-total>{{ renderTotal(total, range) }}</ng-template>\n</nz-table>\n<nz-dropdown-menu #contextmenuTpl=\"nzDropdownMenu\">\n  <ul nz-menu class=\"st__contextmenu\">\n    <ng-container *ngFor=\"let i of contextmenuList\">\n      <li nz-menu-item *ngIf=\"i.children!.length === 0\" (click)=\"i.fn!(i)\" [innerHTML]=\"i.text\"></li>\n      <li nz-submenu *ngIf=\"i.children!.length > 0\" [nzTitle]=\"i.text\">\n        <ul>\n          <li nz-menu-item *ngFor=\"let ci of i.children\" (click)=\"ci.fn!(ci)\" [innerHTML]=\"ci.text\"></li>\n        </ul>\n      </li>\n    </ng-container>\n  </ul>\n</nz-dropdown-menu>\n",
                providers: [STDataSource, STRowSource, STColumnSource, STExport, DatePipe, YNPipe, DecimalPipe],
                host: {
                    '[class.st]': `true`,
                    '[class.st__p-left]': `page.placement === 'left'`,
                    '[class.st__p-center]': `page.placement === 'center'`,
                    '[class.st__width-strict]': `widthMode.type === 'strict'`,
                    '[class.ant-table-rep]': `responsive`,
                    '[class.ant-table-rep__hide-header-footer]': `responsiveHideHeaderFooter`
                },
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [ALAIN_I18N_TOKEN]
            }] }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.ElementRef }, { type: STExport }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: STColumnSource }, { type: STDataSource }, { type: ɵngcc3.DelonLocaleService }, { type: ɵngcc7.AlainConfigService }, { type: ɵngcc8.NzContextMenuService }]; }, { columns: [{
            type: Input
        }], ps: [{
            type: Input
        }], pi: [{
            type: Input
        }], total: [{
            type: Input
        }], loading: [{
            type: Input
        }], loadingDelay: [{
            type: Input
        }], bordered: [{
            type: Input
        }], showHeader: [{
            type: Input
        }], expandRowByClick: [{
            type: Input
        }], expandAccordion: [{
            type: Input
        }], responsive: [{
            type: Input
        }], error: [{
            type: Output
        }], change: [{
            type: Output
        }], virtualScroll: [{
            type: Input
        }], virtualItemSize: [{
            type: Input
        }], virtualMaxBufferPx: [{
            type: Input
        }], virtualMinBufferPx: [{
            type: Input
        }], virtualForTrackBy: [{
            type: Input
        }], req: [{
            type: Input
        }], res: [{
            type: Input
        }], page: [{
            type: Input
        }], multiSort: [{
            type: Input
        }], widthMode: [{
            type: Input
        }], widthConfig: [{
            type: Input
        }], resizable: [{
            type: Input
        }], orgTable: [{
            type: ViewChild,
            args: ['table']
        }], contextmenuTpl: [{
            type: ViewChild,
            args: ['contextmenuTpl']
        }], data: [{
            type: Input
        }], contextmenu: [{
            type: Input
        }], loadingIndicator: [{
            type: Input
        }], size: [{
            type: Input
        }], scroll: [{
            type: Input
        }], singleSort: [{
            type: Input
        }], rowClassName: [{
            type: Input
        }], clickRowClassName: [{
            type: Input
        }], header: [{
            type: Input
        }], footer: [{
            type: Input
        }], bodyHeader: [{
            type: Input
        }], body: [{
            type: Input
        }], expand: [{
            type: Input
        }], noResult: [{
            type: Input
        }], responsiveHideHeaderFooter: [{
            type: Input
        }], customRequest: [{
            type: Input
        }] }); })();

class STFilterComponent {
    constructor() {
        this.visible = false;
        this.locale = {};
        this.n = new EventEmitter();
        this.handle = new EventEmitter();
    }
    get icon() {
        return this.f.icon;
    }
    show($event) {
        $event.stopPropagation();
    }
    checkboxChange() {
        var _a;
        this.n.emit((_a = this.f.menus) === null || _a === void 0 ? void 0 : _a.filter(w => w.checked));
    }
    radioChange(item) {
        this.f.menus.forEach(i => (i.checked = false));
        item.checked = !item.checked;
        this.n.emit(item);
    }
    confirm() {
        this.handle.emit(true);
    }
    reset() {
        this.handle.emit(false);
    }
}
STFilterComponent.ɵfac = function STFilterComponent_Factory(t) { return new (t || STFilterComponent)(); };
STFilterComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: STFilterComponent, selectors: [["st-filter"]], hostVars: 6, hostBindings: function STFilterComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("ant-table-filter-trigger-container", true)("st__filter", true)("ant-table-filter-trigger-container-open", ctx.visible);
    } }, inputs: { locale: "locale", col: "col", f: "f" }, outputs: { n: "n", handle: "handle" }, decls: 13, vars: 14, consts: [["nz-dropdown", "", "nzTrigger", "click", "nzOverlayClassName", "st__filter-wrap", 1, "ant-table-filter-trigger", 3, "nzDropdownMenu", "nzClickHide", "nzVisible", "nzVisibleChange", "click"], ["nz-icon", "", 3, "nzType", "nzTheme"], ["filterMenu", "nzDropdownMenu"], [1, "ant-table-filter-dropdown"], [3, "ngSwitch"], ["class", "st__filter-keyword", 4, "ngSwitchCase"], ["class", "p-sm st__filter-number", 4, "ngSwitchCase"], ["class", "p-sm st__filter-date", 4, "ngSwitchCase"], ["class", "p-sm st__filter-time", 4, "ngSwitchCase"], ["class", "st__filter-custom", 4, "ngSwitchCase"], ["nz-menu", "", 4, "ngSwitchDefault"], ["class", "ant-table-filter-dropdown-btns", 4, "ngIf"], [1, "st__filter-keyword"], ["type", "text", "nz-input", "", 3, "ngModel", "ngModelChange", "keyup.enter"], [1, "p-sm", "st__filter-number"], [1, "width-100", 3, "ngModel", "nzMin", "nzMax", "nzStep", "nzPrecision", "nzPlaceHolder", "ngModelChange"], [1, "p-sm", "st__filter-date"], ["nzInline", "", 3, "nzMode", "ngModel", "nzShowNow", "nzShowToday", "nzDisabledDate", "nzDisabledTime", "ngModelChange", 4, "ngIf"], ["nzInline", "", 3, "nzMode", "ngModel", "nzShowNow", "nzShowToday", "nzDisabledDate", "nzDisabledTime", "ngModelChange"], [1, "p-sm", "st__filter-time"], [1, "st__filter-custom"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["nz-menu", ""], [4, "ngIf"], ["nz-menu-item", "", 4, "ngFor", "ngForOf"], ["nz-menu-item", ""], ["nz-checkbox", "", 3, "ngModel", "ngModelChange"], ["nz-radio", "", 3, "ngModel", "ngModelChange"], [1, "ant-table-filter-dropdown-btns"], [1, "ant-table-filter-dropdown-link", "confirm", 3, "click"], [3, "click"], [1, "ant-table-filter-dropdown-link", "clear", 3, "click"]], template: function STFilterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "span", 0);
        ɵngcc0.ɵɵlistener("nzVisibleChange", function STFilterComponent_Template_span_nzVisibleChange_0_listener($event) { return ctx.visible = $event; })("click", function STFilterComponent_Template_span_click_0_listener($event) { return ctx.show($event); });
        ɵngcc0.ɵɵelement(1, "i", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(2, "nz-dropdown-menu", null, 2);
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementContainerStart(5, 4);
        ɵngcc0.ɵɵtemplate(6, STFilterComponent_div_6_Template, 2, 2, "div", 5);
        ɵngcc0.ɵɵtemplate(7, STFilterComponent_div_7_Template, 2, 6, "div", 6);
        ɵngcc0.ɵɵtemplate(8, STFilterComponent_div_8_Template, 3, 2, "div", 7);
        ɵngcc0.ɵɵtemplate(9, STFilterComponent_div_9_Template, 1, 0, "div", 8);
        ɵngcc0.ɵɵtemplate(10, STFilterComponent_div_10_Template, 2, 5, "div", 9);
        ɵngcc0.ɵɵtemplate(11, STFilterComponent_ul_11_Template, 3, 2, "ul", 10);
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵtemplate(12, STFilterComponent_div_12_Template, 7, 2, "div", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(3);
        ɵngcc0.ɵɵclassProp("active", ctx.visible || ctx.f.default);
        ɵngcc0.ɵɵproperty("nzDropdownMenu", _r0)("nzClickHide", false)("nzVisible", ctx.visible);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("nzType", ctx.icon.type)("nzTheme", ctx.icon.theme);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngSwitch", ctx.f.type);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "keyword");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "number");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "date");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "time");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "custom");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.f.showOPArea);
    } }, directives: [ɵngcc8.NzDropDownDirective, ɵngcc12.NzIconDirective, ɵngcc8.NzDropdownMenuComponent, ɵngcc4.NgSwitch, ɵngcc4.NgSwitchCase, ɵngcc4.NgSwitchDefault, ɵngcc4.NgIf, ɵngcc17.NzInputDirective, ɵngcc14.DefaultValueAccessor, ɵngcc14.NgControlStatus, ɵngcc14.NgModel, ɵngcc18.NzInputNumberComponent, ɵngcc19.NzDatePickerComponent, ɵngcc19.NzRangePickerComponent, ɵngcc4.NgTemplateOutlet, ɵngcc10.NzMenuDirective, ɵngcc4.NgForOf, ɵngcc10.NzMenuItemDirective, ɵngcc13.NzCheckboxComponent, ɵngcc20.NzRadioComponent], encapsulation: 2, changeDetection: 0 });
STFilterComponent.propDecorators = {
    col: [{ type: Input }],
    locale: [{ type: Input }],
    f: [{ type: Input }],
    n: [{ type: Output }],
    handle: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STFilterComponent, [{
        type: Component,
        args: [{
                selector: 'st-filter',
                template: `
    <span
      class="ant-table-filter-trigger"
      [class.active]="visible || f.default"
      nz-dropdown
      [nzDropdownMenu]="filterMenu"
      nzTrigger="click"
      [nzClickHide]="false"
      [(nzVisible)]="visible"
      nzOverlayClassName="st__filter-wrap"
      (click)="show($event)"
    >
      <i nz-icon [nzType]="icon.type" [nzTheme]="icon.theme!"></i>
    </span>
    <nz-dropdown-menu #filterMenu="nzDropdownMenu">
      <div class="ant-table-filter-dropdown">
        <ng-container [ngSwitch]="f.type">
          <div *ngSwitchCase="'keyword'" class="st__filter-keyword">
            <input
              type="text"
              nz-input
              [attr.placeholder]="f.placeholder"
              [(ngModel)]="f.menus![0]!.value"
              (ngModelChange)="n.emit($event)"
              (keyup.enter)="confirm()"
            />
          </div>
          <div *ngSwitchCase="'number'" class="p-sm st__filter-number">
            <nz-input-number
              [(ngModel)]="f.menus![0]!.value"
              (ngModelChange)="n.emit($event)"
              [nzMin]="f.number!.min!"
              [nzMax]="f.number!.max!"
              [nzStep]="f.number!.step!"
              [nzPrecision]="f.number!.precision"
              [nzPlaceHolder]="f.placeholder!"
              class="width-100"
            ></nz-input-number>
          </div>
          <div *ngSwitchCase="'date'" class="p-sm st__filter-date">
            <nz-date-picker
              *ngIf="!f.date!.range"
              nzInline
              [nzMode]="f.date!.mode"
              [(ngModel)]="f.menus![0]!.value"
              (ngModelChange)="n.emit($event)"
              [nzShowNow]="f.date!.showNow"
              [nzShowToday]="f.date!.showToday"
              [nzDisabledDate]="f.date!.disabledDate"
              [nzDisabledTime]="f.date!.disabledTime"
            ></nz-date-picker>
            <nz-range-picker
              *ngIf="f.date!.range"
              nzInline
              [nzMode]="f.date!.mode"
              [(ngModel)]="f.menus![0]!.value"
              (ngModelChange)="n.emit($event)"
              [nzShowNow]="f.date!.showNow"
              [nzShowToday]="f.date!.showToday"
              [nzDisabledDate]="f.date!.disabledDate"
              [nzDisabledTime]="f.date!.disabledTime"
            ></nz-range-picker>
          </div>
          <div *ngSwitchCase="'time'" class="p-sm st__filter-time"> </div>
          <div *ngSwitchCase="'custom'" class="st__filter-custom">
            <ng-template
              [ngTemplateOutlet]="f.custom!"
              [ngTemplateOutletContext]="{ $implicit: f, col: col }"
            ></ng-template>
          </div>
          <ul *ngSwitchDefault nz-menu>
            <ng-container *ngIf="f.multiple">
              <li nz-menu-item *ngFor="let filter of f.menus">
                <label nz-checkbox [(ngModel)]="filter.checked" (ngModelChange)="checkboxChange()">
                  {{ filter.text }}
                </label>
              </li>
            </ng-container>
            <ng-container *ngIf="!f.multiple">
              <li nz-menu-item *ngFor="let filter of f.menus">
                <label nz-radio [ngModel]="filter.checked" (ngModelChange)="radioChange(filter)">
                  {{ filter.text }}
                </label>
              </li>
            </ng-container>
          </ul>
        </ng-container>
        <div *ngIf="f.showOPArea" class="ant-table-filter-dropdown-btns">
          <a class="ant-table-filter-dropdown-link confirm" (click)="visible = false">
            <span (click)="confirm()">{{ f.confirmText || locale.filterConfirm }}</span>
          </a>
          <a class="ant-table-filter-dropdown-link clear" (click)="visible = false">
            <span (click)="reset()">{{ f.clearText || locale.filterReset }}</span>
          </a>
        </div>
      </div>
    </nz-dropdown-menu>
  `,
                host: {
                    '[class.ant-table-filter-trigger-container]': `true`,
                    '[class.st__filter]': `true`,
                    '[class.ant-table-filter-trigger-container-open]': `visible`
                },
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return []; }, { locale: [{
            type: Input
        }], n: [{
            type: Output
        }], handle: [{
            type: Output
        }], col: [{
            type: Input
        }], f: [{
            type: Input
        }] }); })();

class STTdComponent {
    constructor(stComp, router, modalHelper, drawerHelper) {
        this.stComp = stComp;
        this.router = router;
        this.modalHelper = modalHelper;
        this.drawerHelper = drawerHelper;
        this.n = new EventEmitter();
    }
    get routerState() {
        const { pi, ps, total } = this.stComp;
        return { pi, ps, total };
    }
    report(type) {
        this.n.emit({ type, item: this.i, col: this.c });
    }
    _checkbox(value) {
        this.i.checked = value;
        this.report('checkbox');
    }
    _radio(checked) {
        this.data.filter(w => !w.disabled).forEach(i => (i.checked = false));
        this.i.checked = checked;
        this.report('radio');
    }
    _link(e) {
        this._stopPropagation(e);
        const res = this.c.click(this.i, this.stComp);
        if (typeof res === 'string') {
            this.router.navigateByUrl(res, { state: this.routerState });
        }
        return false;
    }
    _stopPropagation(ev) {
        ev.preventDefault();
        ev.stopPropagation();
    }
    _btn(btn, ev) {
        if (ev) {
            ev.stopPropagation();
        }
        const record = this.i;
        if (btn.type === 'modal' || btn.type === 'static') {
            const { modal } = btn;
            const obj = { [modal.paramsName]: record };
            this.modalHelper[btn.type === 'modal' ? 'create' : 'createStatic'](modal.component, Object.assign(Object.assign({}, obj), (modal.params && modal.params(record))), deepMergeKey({}, true, this.stComp['cog'].modal, modal))
                .pipe(filter(w => typeof w !== 'undefined'))
                .subscribe((res) => this.btnCallback(record, btn, res));
            return;
        }
        else if (btn.type === 'drawer') {
            const { drawer } = btn;
            const obj = { [drawer.paramsName]: record };
            this.drawerHelper
                .create(drawer.title, drawer.component, Object.assign(Object.assign({}, obj), (drawer.params && drawer.params(record))), deepMergeKey({}, true, this.stComp['cog'].drawer, drawer))
                .pipe(filter(w => typeof w !== 'undefined'))
                .subscribe(res => this.btnCallback(record, btn, res));
            return;
        }
        else if (btn.type === 'link') {
            const clickRes = this.btnCallback(record, btn);
            if (typeof clickRes === 'string') {
                this.router.navigateByUrl(clickRes, { state: this.routerState });
            }
            return;
        }
        this.btnCallback(record, btn);
    }
    btnCallback(record, btn, modal) {
        if (!btn.click)
            return;
        if (typeof btn.click === 'string') {
            switch (btn.click) {
                case 'load':
                    this.stComp.load();
                    break;
                case 'reload':
                    this.stComp.reload();
                    break;
            }
        }
        else {
            return btn.click(record, modal, this.stComp);
        }
    }
}
STTdComponent.ɵfac = function STTdComponent_Factory(t) { return new (t || STTdComponent)(ɵngcc0.ɵɵdirectiveInject(STComponent, 1), ɵngcc0.ɵɵdirectiveInject(ɵngcc21.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ModalHelper), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DrawerHelper)); };
STTdComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: STTdComponent, selectors: [["st-td"]], inputs: { c: "c", cIdx: "cIdx", data: "data", i: "i", index: "index" }, outputs: { n: "n" }, decls: 9, vars: 8, consts: [["btnTpl", ""], ["btnItemTpl", ""], ["btnTextTpl", ""], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["render", ""], [4, "ngIf", "ngIfElse"], [4, "ngIf"], ["nz-tooltip", "", 3, "nzTooltipTitle", 4, "ngIf"], ["nz-tooltip", "", 3, "nzTooltipTitle"], ["nz-popconfirm", "", "class", "st__btn-text", 3, "nzPopconfirmTitle", "nzIcon", "nzCondition", "nzCancelText", "nzOkText", "nzOkType", "ngClass", "nzOnConfirm", "click", 4, "ngIf"], ["class", "st__btn-text", 3, "ngClass", "click", 4, "ngIf"], ["nz-popconfirm", "", 1, "st__btn-text", 3, "nzPopconfirmTitle", "nzIcon", "nzCondition", "nzCancelText", "nzOkText", "nzOkType", "ngClass", "nzOnConfirm", "click"], [1, "st__btn-text", 3, "ngClass", "click"], [3, "innerHTML", "ngClass"], ["nz-icon", "", 3, "nzType", "nzTheme", "nzSpin", "nzTwotoneColor", 4, "ngIf"], ["nz-icon", "", 3, "nzIconfont", 4, "ngIf"], ["nz-icon", "", 3, "nzType", "nzTheme", "nzSpin", "nzTwotoneColor"], ["nz-icon", "", 3, "nzIconfont"], [3, "ngSwitch"], ["nz-checkbox", "", 3, "nzDisabled", "ngModel", "ngModelChange", 4, "ngSwitchCase"], ["nz-radio", "", 3, "nzDisabled", "ngModel", "ngModelChange", 4, "ngSwitchCase"], [3, "innerHTML", "click", 4, "ngSwitchCase"], [4, "ngSwitchCase"], [4, "ngSwitchDefault"], [4, "ngFor", "ngForOf"], ["nz-checkbox", "", 3, "nzDisabled", "ngModel", "ngModelChange"], ["nz-radio", "", 3, "nzDisabled", "ngModel", "ngModelChange"], [3, "innerHTML", "click"], [3, "nzColor", 4, "ngSwitchCase"], [3, "nzStatus", "nzText", 4, "ngSwitchCase"], [3, "nzColor"], [3, "innerHTML"], [3, "nzStatus", "nzText"], ["st-widget-host", "", 3, "record", "column"], [3, "innerHTML", 4, "ngIf"], [3, "innerText", 4, "ngIf"], [3, "innerText"], ["nz-dropdown", "", "nzOverlayClassName", "st__btn-sub", 3, "nzDropdownMenu", 4, "ngIf"], ["btnMenu", "nzDropdownMenu"], ["nz-menu", ""], [3, "st__btn-disabled", 4, "ngIf"], ["nzType", "vertical", 4, "ngIf"], ["nz-dropdown", "", "nzOverlayClassName", "st__btn-sub", 3, "nzDropdownMenu"], ["nz-icon", "", "nzType", "down"], ["nz-menu-item", "", 3, "st__btn-disabled", 4, "ngIf"], ["nz-menu-divider", "", 4, "ngIf"], ["nz-menu-item", ""], ["nz-menu-divider", ""], ["nzType", "vertical"]], template: function STTdComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, STTdComponent_ng_template_0_Template, 2, 2, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(2, STTdComponent_ng_template_2_Template, 2, 2, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(4, STTdComponent_ng_template_4_Template, 2, 5, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(6, STTdComponent_ng_template_6_Template, 0, 0, "ng-template", 3, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(8, STTdComponent_ng_container_8_Template, 9, 7, "ng-container", 5);
    } if (rf & 2) {
        const _r6 = ɵngcc0.ɵɵreference(7);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.c.__render)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(4, _c8, ctx.i, ctx.index, ctx.c));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.c.__render)("ngIfElse", _r6);
    } }, directives: [ɵngcc4.NgTemplateOutlet, ɵngcc4.NgIf, ɵngcc11.NzTooltipDirective, ɵngcc22.NzPopconfirmDirective, ɵngcc4.NgClass, ɵngcc12.NzIconDirective, ɵngcc4.NgSwitch, ɵngcc4.NgSwitchCase, ɵngcc4.NgSwitchDefault, ɵngcc4.NgForOf, ɵngcc13.NzCheckboxComponent, ɵngcc14.NgControlStatus, ɵngcc14.NgModel, ɵngcc20.NzRadioComponent, ɵngcc23.NzTagComponent, ɵngcc24.NzBadgeComponent, STWidgetHostDirective, ɵngcc8.NzDropdownMenuComponent, ɵngcc10.NzMenuDirective, ɵngcc8.NzDropDownADirective, ɵngcc8.NzDropDownDirective, ɵngcc10.NzMenuItemDirective, ɵngcc10.NzMenuDividerDirective, ɵngcc25.NzDividerComponent], encapsulation: 2, changeDetection: 0 });
STTdComponent.ctorParameters = () => [
    { type: STComponent, decorators: [{ type: Host }] },
    { type: Router },
    { type: ModalHelper },
    { type: DrawerHelper }
];
STTdComponent.propDecorators = {
    c: [{ type: Input }],
    cIdx: [{ type: Input }],
    data: [{ type: Input }],
    i: [{ type: Input }],
    index: [{ type: Input }],
    n: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STTdComponent, [{
        type: Component,
        args: [{
                selector: 'st-td',
                template: `
    <ng-template #btnTpl let-i>
      <ng-container *ngIf="!i.tooltip">
        <ng-template [ngTemplateOutlet]="btnItemTpl" [ngTemplateOutletContext]="{ $implicit: i }"></ng-template>
      </ng-container>
      <span *ngIf="i.tooltip" nz-tooltip [nzTooltipTitle]="i.tooltip">
        <ng-template [ngTemplateOutlet]="btnItemTpl" [ngTemplateOutletContext]="{ $implicit: i }"></ng-template>
      </span>
    </ng-template>
    <ng-template #btnItemTpl let-i>
      <a
        *ngIf="i.pop"
        nz-popconfirm
        [nzPopconfirmTitle]="i.pop.title"
        [nzIcon]="i.pop.icon"
        [nzCondition]="i.pop.condition(i)"
        [nzCancelText]="i.pop.cancelText"
        [nzOkText]="i.pop.okText"
        [nzOkType]="i.pop.okType"
        (nzOnConfirm)="_btn(i)"
        class="st__btn-text"
        [ngClass]="i.className"
        (click)="_stopPropagation($event)"
      >
        <ng-template [ngTemplateOutlet]="btnTextTpl" [ngTemplateOutletContext]="{ $implicit: i }"></ng-template>
      </a>
      <a *ngIf="!i.pop" (click)="_btn(i, $event)" class="st__btn-text" [ngClass]="i.className">
        <ng-template [ngTemplateOutlet]="btnTextTpl" [ngTemplateOutletContext]="{ $implicit: i }"></ng-template>
      </a>
    </ng-template>
    <ng-template #btnTextTpl let-i>
      <ng-container *ngIf="i.icon">
        <i
          *ngIf="!i.icon.iconfont"
          nz-icon
          [nzType]="i.icon.type"
          [nzTheme]="i.icon.theme"
          [nzSpin]="i.icon.spin"
          [nzTwotoneColor]="i.icon.twoToneColor"
        ></i>
        <i *ngIf="i.icon.iconfont" nz-icon [nzIconfont]="i.icon.iconfont"></i>
      </ng-container>
      <span [innerHTML]="i._text" [ngClass]="{ 'pl-xs': i.icon }"></span>
    </ng-template>
    <ng-template
      #render
      [ngTemplateOutlet]="c.__render!"
      [ngTemplateOutletContext]="{ $implicit: i, index: index, column: c }"
    ></ng-template>
    <ng-container *ngIf="!c.__render; else render">
      <ng-container [ngSwitch]="c.type">
        <label
          *ngSwitchCase="'checkbox'"
          nz-checkbox
          [nzDisabled]="i.disabled"
          [ngModel]="i.checked"
          (ngModelChange)="_checkbox($event)"
        ></label>
        <label
          *ngSwitchCase="'radio'"
          nz-radio
          [nzDisabled]="i.disabled"
          [ngModel]="i.checked"
          (ngModelChange)="_radio($event)"
        ></label>
        <a
          *ngSwitchCase="'link'"
          (click)="_link($event)"
          [innerHTML]="i._values[cIdx]._text"
          [attr.title]="i._values[cIdx].text"
        ></a>
        <ng-container *ngIf="i._values[cIdx].text">
          <nz-tag *ngSwitchCase="'tag'" [nzColor]="i._values[cIdx].color">
            <span [innerHTML]="i._values[cIdx]._text"></span>
          </nz-tag>
          <nz-badge
            *ngSwitchCase="'badge'"
            [nzStatus]="i._values[cIdx].color"
            [nzText]="i._values[cIdx].text"
          ></nz-badge>
        </ng-container>
        <ng-template *ngSwitchCase="'widget'" st-widget-host [record]="i" [column]="c"></ng-template
        ><ng-container *ngSwitchDefault>
          <span
            *ngIf="c.safeType !== 'text'"
            [innerHTML]="i._values[cIdx]._text"
            [attr.title]="c._isTruncate ? i._values[cIdx].text : null"
          ></span>
          <span
            *ngIf="c.safeType === 'text'"
            [innerText]="i._values[cIdx]._text"
            [attr.title]="c._isTruncate ? i._values[cIdx].text : null"
          ></span>
        </ng-container>
      </ng-container>
      <ng-container *ngFor="let btn of i._values[cIdx].buttons; let last = last">
        <a *ngIf="btn.children!.length > 0" nz-dropdown [nzDropdownMenu]="btnMenu" nzOverlayClassName="st__btn-sub">
          <span [innerHTML]="btn._text"></span>
          <i nz-icon nzType="down"></i>
        </a>
        <nz-dropdown-menu #btnMenu="nzDropdownMenu">
          <ul nz-menu>
            <ng-container *ngFor="let subBtn of btn.children!">
              <li *ngIf="subBtn.type !== 'divider'" nz-menu-item [class.st__btn-disabled]="subBtn._disabled">
                <ng-template [ngTemplateOutlet]="btnTpl" [ngTemplateOutletContext]="{ $implicit: subBtn }">
                </ng-template>
              </li>
              <li *ngIf="subBtn.type === 'divider'" nz-menu-divider></li>
            </ng-container>
          </ul>
        </nz-dropdown-menu>
        <span *ngIf="btn.children!.length === 0" [class.st__btn-disabled]="btn._disabled">
          <ng-template [ngTemplateOutlet]="btnTpl" [ngTemplateOutletContext]="{ $implicit: btn }"> </ng-template>
        </span>
        <nz-divider *ngIf="!last" nzType="vertical"></nz-divider>
      </ng-container>
    </ng-container>
  `,
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: STComponent, decorators: [{
                type: Host
            }] }, { type: ɵngcc21.Router }, { type: ɵngcc3.ModalHelper }, { type: ɵngcc3.DrawerHelper }]; }, { n: [{
            type: Output
        }], c: [{
            type: Input
        }], cIdx: [{
            type: Input
        }], data: [{
            type: Input
        }], i: [{
            type: Input
        }], index: [{
            type: Input
        }] }); })();

const COMPONENTS = [STComponent, STRowDirective, STWidgetHostDirective];
class STModule {
}
STModule.ɵfac = function STModule_Factory(t) { return new (t || STModule)(); };
STModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: STModule });
STModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            DelonACLModule,
            LetModule,
            NzPopconfirmModule,
            NzTableModule,
            NzIconModule,
            NzBadgeModule,
            NzCheckboxModule,
            NzDividerModule,
            NzDropDownModule,
            NzMenuModule,
            NzRadioModule,
            NzTagModule,
            NzInputModule,
            NzToolTipModule,
            NzResizableModule,
            NzInputNumberModule,
            NzDatePickerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(STModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    DelonACLModule,
                    LetModule,
                    NzPopconfirmModule,
                    NzTableModule,
                    NzIconModule,
                    NzBadgeModule,
                    NzCheckboxModule,
                    NzDividerModule,
                    NzDropDownModule,
                    NzMenuModule,
                    NzRadioModule,
                    NzTagModule,
                    NzInputModule,
                    NzToolTipModule,
                    NzResizableModule,
                    NzInputNumberModule,
                    NzDatePickerModule
                ],
                declarations: [...COMPONENTS, STFilterComponent, STTdComponent],
                exports: COMPONENTS
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(STModule, { declarations: function () { return [STComponent, STRowDirective, STWidgetHostDirective, STFilterComponent, STTdComponent]; }, imports: function () { return [CommonModule,
        FormsModule,
        DelonACLModule,
        LetModule,
        NzPopconfirmModule,
        NzTableModule,
        NzIconModule,
        NzBadgeModule,
        NzCheckboxModule,
        NzDividerModule,
        NzDropDownModule,
        NzMenuModule,
        NzRadioModule,
        NzTagModule,
        NzInputModule,
        NzToolTipModule,
        NzResizableModule,
        NzInputNumberModule,
        NzDatePickerModule]; }, exports: function () { return [STComponent, STRowDirective, STWidgetHostDirective]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { STColumnSource, STComponent, STDataSource, STExport, STModule, STRowDirective, STWidgetHostDirective, STWidgetRegistry, ST_DEFAULT_CONFIG, STRowSource as ɵa, STFilterComponent as ɵb, STTdComponent as ɵc };
